// @turnkey/sdk-browser/dist/storage.js
'use strict';

var window = require('./__polyfills__/window.js');

exports.StorageKeys = void 0;
(function (StorageKeys) {
    StorageKeys["CurrentUser"] = "@turnkey/current_user";
})(exports.StorageKeys || (exports.StorageKeys = {}));
var StorageLocation;
(function (StorageLocation) {
    StorageLocation["Local"] = "local";
    StorageLocation["Secure"] = "secure";
    StorageLocation["Session"] = "session";
})(StorageLocation || (StorageLocation = {}));
const STORAGE_VALUE_LOCATIONS = {
    [exports.StorageKeys.CurrentUser]: StorageLocation.Local,
};
const STORAGE_LOCATIONS = {
    [StorageLocation.Local]: window.localStorage,
    [StorageLocation.Secure]: window.localStorage,
    [StorageLocation.Session]: window.localStorage,
};
const getStorageValue = async (storageKey) => {
    const storageLocation = STORAGE_VALUE_LOCATIONS[storageKey];
    const browserStorageLocation = STORAGE_LOCATIONS[storageLocation];
    const storageItem = browserStorageLocation.getItem(storageKey);
    return storageItem ? JSON.parse(storageItem) : undefined;
};
const setStorageValue = async (storageKey, storageValue) => {
    const storageLocation = STORAGE_VALUE_LOCATIONS[storageKey];
    const browserStorageLocation = STORAGE_LOCATIONS[storageLocation];
    browserStorageLocation.setItem(storageKey, JSON.stringify(storageValue));
};
const removeStorageValue = async (storageKey) => {
    const storageLocation = STORAGE_VALUE_LOCATIONS[storageKey];
    const browserStorageLocation = STORAGE_LOCATIONS[storageLocation];
    browserStorageLocation.removeItem(storageKey);
};

exports.getStorageValue = getStorageValue;
exports.removeStorageValue = removeStorageValue;
exports.setStorageValue = setStorageValue;
