// @alchemy/aa-core/dist/cjs/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.Logger = exports.resolveProperties = exports.getUserOperationHash = exports.getChain = exports.defineReadOnly = exports.deepHexlify = exports.asyncPipe = exports.noOpMiddleware = exports.SmartAccountProvider = exports.createPublicErc4337FromClient = exports.createPublicErc4337Client = exports.wrapWith6492 = exports.PrivateKeySigner = exports.LocalAccountSigner = exports.HdAccountSigner = exports.SimpleSmartContractAccount = exports.BaseSmartContractAccount = exports.SimpleAccountFactoryAbi = exports.SimpleAccountAbi = exports.EntryPointAbi = exports.chains = void 0;
exports.chains = __importStar(require("viem/chains"));
var EntryPointAbi_js_1 = require("./abis/EntryPointAbi.js");
Object.defineProperty(exports, "EntryPointAbi", { enumerable: true, get: function () { return EntryPointAbi_js_1.EntryPointAbi; } });
var SimpleAccountAbi_js_1 = require("./abis/SimpleAccountAbi.js");
Object.defineProperty(exports, "SimpleAccountAbi", { enumerable: true, get: function () { return SimpleAccountAbi_js_1.SimpleAccountAbi; } });
var SimpleAccountFactoryAbi_js_1 = require("./abis/SimpleAccountFactoryAbi.js");
Object.defineProperty(exports, "SimpleAccountFactoryAbi", { enumerable: true, get: function () { return SimpleAccountFactoryAbi_js_1.SimpleAccountFactoryAbi; } });
var base_js_1 = require("./account/base.js");
Object.defineProperty(exports, "BaseSmartContractAccount", { enumerable: true, get: function () { return base_js_1.BaseSmartContractAccount; } });
var simple_js_1 = require("./account/simple.js");
Object.defineProperty(exports, "SimpleSmartContractAccount", { enumerable: true, get: function () { return simple_js_1.SimpleSmartContractAccount; } });
var hd_account_js_1 = require("./signer/hd-account.js");
Object.defineProperty(exports, "HdAccountSigner", { enumerable: true, get: function () { return hd_account_js_1.HdAccountSigner; } });
var local_account_js_1 = require("./signer/local-account.js");
Object.defineProperty(exports, "LocalAccountSigner", { enumerable: true, get: function () { return local_account_js_1.LocalAccountSigner; } });
var private_key_js_1 = require("./signer/private-key.js");
Object.defineProperty(exports, "PrivateKeySigner", { enumerable: true, get: function () { return private_key_js_1.PrivateKeySigner; } });
var utils_js_1 = require("./signer/utils.js");
Object.defineProperty(exports, "wrapWith6492", { enumerable: true, get: function () { return utils_js_1.wrapWith6492; } });
var create_client_js_1 = require("./client/create-client.js");
Object.defineProperty(exports, "createPublicErc4337Client", { enumerable: true, get: function () { return create_client_js_1.createPublicErc4337Client; } });
Object.defineProperty(exports, "createPublicErc4337FromClient", { enumerable: true, get: function () { return create_client_js_1.createPublicErc4337FromClient; } });
var base_js_2 = require("./provider/base.js");
Object.defineProperty(exports, "SmartAccountProvider", { enumerable: true, get: function () { return base_js_2.SmartAccountProvider; } });
Object.defineProperty(exports, "noOpMiddleware", { enumerable: true, get: function () { return base_js_2.noOpMiddleware; } });
var utils_js_2 = require("./utils.js");
Object.defineProperty(exports, "asyncPipe", { enumerable: true, get: function () { return utils_js_2.asyncPipe; } });
Object.defineProperty(exports, "deepHexlify", { enumerable: true, get: function () { return utils_js_2.deepHexlify; } });
Object.defineProperty(exports, "defineReadOnly", { enumerable: true, get: function () { return utils_js_2.defineReadOnly; } });
Object.defineProperty(exports, "getChain", { enumerable: true, get: function () { return utils_js_2.getChain; } });
Object.defineProperty(exports, "getUserOperationHash", { enumerable: true, get: function () { return utils_js_2.getUserOperationHash; } });
Object.defineProperty(exports, "resolveProperties", { enumerable: true, get: function () { return utils_js_2.resolveProperties; } });
var logger_js_1 = require("./logger.js");
Object.defineProperty(exports, "Logger", { enumerable: true, get: function () { return logger_js_1.Logger; } });
