// @zerodev/sdk/dist/cjs/kernel-zerodev/constants.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.CHAIN_ID_TO_NODE = exports.INFURA_API_KEY = exports.minPriorityFeePerBidDefaults = exports.ERC20_APPROVAL_AMOUNT = exports.gasTokenChainAddresses = exports.oneAddress = exports.RECOVERY_ACTION = exports.RECOVERY_VALIDATOR_ADDRESS = exports.SESSION_KEY_VALIDATOR_ADDRESS = exports.TOKEN_ACTION = exports.ERC165_SESSION_KEY_VALIDATOR_ADDRESS = exports.DUMMY_ECDSA_SIG = exports.KILL_SWITCH_ACTION = exports.KILL_SWITCH_VALIDATOR_ADDRESS = exports.KERNEL_IMPL_ADDRESS = exports.KERNEL_FACTORY_ADDRESS = exports.ECDSA_VALIDATOR_ADDRESS = exports.API_URL = exports.BACKEND_URL = exports.MULTISEND_ADDR = exports.ENTRYPOINT_ADDRESS = exports.PAYMASTER_URL = exports.BUNDLER_URL = exports.DEFAULT_SEND_TX_RETRY_INTERVAL_MS = exports.DEFAULT_SEND_TX_MAX_RETRIES = void 0;
const viem_1 = require("viem");
const chains_1 = require("viem/chains");
exports.DEFAULT_SEND_TX_MAX_RETRIES = 3;
exports.DEFAULT_SEND_TX_RETRY_INTERVAL_MS = 60000;
exports.BUNDLER_URL = "https://meta-aa-provider.onrender.com/api/v1/bundler";
exports.PAYMASTER_URL = "https://meta-aa-provider.onrender.com/api/v1/paymaster";
exports.ENTRYPOINT_ADDRESS = "0x5FF137D4b0FDCD49DcA30c7CF57E578a026d2789";
exports.MULTISEND_ADDR = "0x8ae01fcf7c655655ff2c6ef907b8b4718ab4e17c";
exports.BACKEND_URL = "https://backend-vikp.onrender.com";
exports.API_URL = "https://prod-api.zerodev.app";
exports.ECDSA_VALIDATOR_ADDRESS = "0xd9AB5096a832b9ce79914329DAEE236f8Eea0390";
exports.KERNEL_FACTORY_ADDRESS = "0x5de4839a76cf55d0c90e2061ef4386d962E15ae3";
exports.KERNEL_IMPL_ADDRESS = "0x0DA6a956B9488eD4dd761E59f52FDc6c8068E6B5";
exports.KILL_SWITCH_VALIDATOR_ADDRESS = "0x7393A7dA58CCfFb78f52adb09705BE6E20F704BC";
exports.KILL_SWITCH_ACTION = "0x3f38e479304c7F18F988269a1bDa7d646bd48243";
exports.DUMMY_ECDSA_SIG = "0xfffffffffffffffffffffffffffffff0000000000000000000000000000000007aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa1c";
exports.ERC165_SESSION_KEY_VALIDATOR_ADDRESS = "0xe149290800De29D4b0BF9dB82c508255D81902E6";
exports.TOKEN_ACTION = "0x2087C7FfD0d0DAE80a00EE74325aBF3449e0eaf1";
exports.SESSION_KEY_VALIDATOR_ADDRESS = "0x5C06CE2b673fD5E6e56076e40DD46aB67f5a72A5";
exports.RECOVERY_VALIDATOR_ADDRESS = "0xaFA369cA3432e62ACe6bd77B2886F2D588c68dA8";
exports.RECOVERY_ACTION = "0x2f65dB8039fe5CAEE0a8680D2879deB800F31Ae1";
exports.oneAddress = "0x0000000000000000000000000000000000000001";
exports.gasTokenChainAddresses = {
    USDC: {
        1: "0xA0b86991c6218b36c1d19D4a2e9Eb0cE3606eB48",
        137: "0x2791Bca1f2de4661ED88A30C99A7a9449Aa84174",
        42161: "0xFF970A61A04b1cA14834A43f5dE4533eBDDB5CC8",
        43114: "0xB97EF9Ef8734C71904D8002F8b6Bc66Dd9c48a6E",
    },
    PEPE: {
        1: "0x6982508145454Ce325dDbE47a25d4ec3d2311933",
    },
};
exports.ERC20_APPROVAL_AMOUNT = {
    [exports.gasTokenChainAddresses["USDC"][1]]: (0, viem_1.parseUnits)("100", 6),
    [exports.gasTokenChainAddresses["USDC"][5]]: (0, viem_1.parseUnits)("100", 6),
    [exports.gasTokenChainAddresses["USDC"][137]]: (0, viem_1.parseUnits)("10", 6),
    [exports.gasTokenChainAddresses["USDC"][42161]]: (0, viem_1.parseUnits)("10", 6),
    [exports.gasTokenChainAddresses["USDC"][43114]]: (0, viem_1.parseUnits)("10", 6),
    [exports.gasTokenChainAddresses["PEPE"][1]]: (0, viem_1.parseUnits)("50000000", 18),
    "0x3870419Ba2BBf0127060bCB37f69A1b1C090992B": (0, viem_1.parseUnits)("100", 18),
};
exports.minPriorityFeePerBidDefaults = new Map([
    [chains_1.polygon.id, 30000000000n],
]);
exports.INFURA_API_KEY = "f36f7f706a58477884ce6fe89165666c";
exports.CHAIN_ID_TO_NODE = {
    1: `https://mainnet.infura.io/v3/${exports.INFURA_API_KEY}`,
    5: `https://goerli.infura.io/v3/${exports.INFURA_API_KEY}`,
    11155111: "https://fittest-ultra-aura.ethereum-sepolia.quiknode.pro/3893d01b1dd411fdfa9b6dd372dd2b4f69fcf1ea/",
    137: `https://polygon-mainnet.infura.io/v3/${exports.INFURA_API_KEY}`,
    80001: `https://polygon-mumbai.infura.io/v3/${exports.INFURA_API_KEY}`,
    80002: `https://polygon-amoy.infura.io/v3/${exports.INFURA_API_KEY}`,
    10: `https://optimism-mainnet.infura.io/v3/${exports.INFURA_API_KEY}`,
    420: `https://optimism-goerli.infura.io/v3/${exports.INFURA_API_KEY}`,
    42161: `https://arbitrum-mainnet.infura.io/v3/${exports.INFURA_API_KEY}`,
    421613: `https://arbitrum-goerli.infura.io/v3/${exports.INFURA_API_KEY}`,
    43114: `https://avalanche-mainnet.infura.io/v3/${exports.INFURA_API_KEY}`,
    43113: `https://avalanche-fuji.infura.io/v3/${exports.INFURA_API_KEY}`,
    1313161554: `https://aurora-mainnet.infura.io/v3/${exports.INFURA_API_KEY}`,
    1313161555: `https://aurora-testnet.infura.io/v3/${exports.INFURA_API_KEY}`,
    56: "https://neat-greatest-layer.bsc.quiknode.pro/9405a499ceee314e5f2f68c9d47518d3537fce6a/",
    8453: "https://twilight-red-tree.base-mainnet.quiknode.pro/dc6eb27bf0f917df215922488dd97f4de7d9b08e/",
    84531: "https://icy-long-mountain.base-goerli.quiknode.pro/5b80d93e97cc9412a63c10a30841869abbef9596/",
    100: "https://thrilling-fluent-film.xdai.quiknode.pro/305955cffb9868cdd95b5e3dc9775f20678ad9ac/",
    10200: "https://nd-810-853-201.p2pify.com/e828b09f0d43591de96c297b3f36fffd",
    59144: `https://linea-mainnet.infura.io/v3/${exports.INFURA_API_KEY}`,
    59140: `https://linea-goerli.infura.io/v3/${exports.INFURA_API_KEY}`,
    204: "https://opbnb.rpc.thirdweb.com/eed98c1f30120c0508f17c1f01d3102b",
    42069: "https://rpc.op-testnet.gelato.digital",
    1261120: "https://rpc.zkatana.gelato.digital",
    84532: "https://floral-old-feather.base-sepolia.quiknode.pro/4c3b4295d0c174410872b1e461a91cc33c7969a9/",
    421614: `https://arbitrum-sepolia.infura.io/v3/${exports.INFURA_API_KEY}`,
    11155420: `https://optimism-sepolia.infura.io/v3/${exports.INFURA_API_KEY}`,
};
