// @zerodev/sdk/dist/cjs/kernel-zerodev/client/create-client.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.createZeroDevPublicErc4337Client = void 0;
const aa_core_1 = require("@alchemy/aa-core");
const viem_1 = require("viem");
const constants_js_1 = require("../constants.js");
function isMethodInErc4337RpcSchema(variable) {
    return [
        "eth_sendUserOperation",
        "eth_estimateUserOperationGas",
        "eth_getUserOperationReceipt",
        "eth_getUserOperationByHash",
        "eth_supportedEntryPoints",
        "eth_maxPriorityFeePerGas",
        "rundler_maxPriorityFeePerGas",
        "pimlico_getUserOperationGasPrice",
    ].includes(variable);
}
const createZeroDevPublicErc4337Client = ({ chain, rpcUrl, bundlerRpcUrl, projectId, bundlerProvider, }) => {
    const erc4337Transport = (0, viem_1.http)(bundlerRpcUrl ?? constants_js_1.BUNDLER_URL, {
        fetchOptions: {
            headers: bundlerRpcUrl === constants_js_1.BUNDLER_URL ? { projectId, bundlerProvider } : {},
        },
        name: "Connected bundler network",
        key: "connected-bundler-network",
        retryCount: 0,
        timeout: 35000,
    });
    const publicTransport = (0, viem_1.http)(rpcUrl ?? constants_js_1.CHAIN_ID_TO_NODE[chain.id]);
    let client = (0, aa_core_1.createPublicErc4337FromClient)((0, viem_1.createPublicClient)({
        chain,
        transport: (0, viem_1.custom)({
            async request({ method, params }) {
                let response;
                if (isMethodInErc4337RpcSchema(method)) {
                    response = await erc4337Transport({ chain }).request({
                        method,
                        params,
                    });
                }
                else {
                    response = await publicTransport({ chain }).request({
                        method,
                        params,
                    });
                }
                return response;
            },
        }),
    }));
    return client;
};
exports.createZeroDevPublicErc4337Client = createZeroDevPublicErc4337Client;
