// @dynamic-labs/sdk-react-core/src/lib/widgets/DynamicWidget/views/ManagePasskeysWidgetView/PasskeyCard/PasskeyCard.cjs
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var jsxRuntime = require('react/jsx-runtime');
var React = require('react');
var reactI18next = require('react-i18next');
var iconic = require('@dynamic-labs/iconic');
var DotsMenu = require('../../../components/DotsMenu/DotsMenu.cjs');
var classNames = require('../../../../../utils/functions/classNames/classNames.cjs');
require('../../../../../events/dynamicEvents.cjs');
require('../../../../../../../_virtual/_tslib.cjs');
require('@dynamic-labs/utils');
require('../../../../../context/DynamicContext/DynamicContext.cjs');
require('@dynamic-labs/sdk-api');
require('../../../../../shared/logger.cjs');
require('@dynamic-labs/wallet-connector-core');
var ViewContext = require('../../../../../context/ViewContext/ViewContext.cjs');
require('@dynamic-labs/wallet-book');
require('../../../../../utils/constants/colors.cjs');
require('../../../../../utils/constants/values.cjs');
require('../../../../../shared/utils/classes/storage/localStorage.cjs');
require('viem');
require('../../../../../shared/consts/index.cjs');
var useInternalDynamicContext = require('../../../../../context/DynamicContext/useDynamicContext/useInternalDynamicContext.cjs');
require('../../../../../context/CaptchaContext/CaptchaContext.cjs');
require('../../../../../context/ErrorContext/ErrorContext.cjs');
require('@dynamic-labs/multi-wallet');
var shortenEmail = require('../../../../../shared/utils/functions/shortenEmail/shortenEmail.cjs');
require('viem/chains');
require('../../../../../context/AccessDeniedContext/AccessDeniedContext.cjs');
require('../../../../../context/AccountExistsContext/AccountExistsContext.cjs');
require('../../../../../config/ApiEndpoint.cjs');
require('../../../../../context/EmailVerificationContext/EmailVerificationContext.cjs');
require('react-dom');
require('../../../../../context/ThemeContext/ThemeContext.cjs');
require('@dynamic-labs/types');
require('yup');
require('../../../../../context/MockContext/MockContext.cjs');
require('../../../../../utils/hooks/useUserUpdateRequest/useUpdateUser/useUpdateUser.cjs');
require('../../../../../context/UserFieldEditorContext/UserFieldEditorContext.cjs');
require('@dynamic-labs/rpc-providers');
require('../../../../../context/UserWalletsContext/UserWalletsContext.cjs');
require('../../../../../components/Transition/ZoomTransition/ZoomTransition.cjs');
require('../../../../../components/Transition/SlideInUpTransition/SlideInUpTransition.cjs');
require('../../../../../components/Transition/OpacityTransition/OpacityTransition.cjs');
require('../../../../../components/ShadowDOM/ShadowDOM.cjs');
require('../../../../../components/OverlayCard/OverlayCard.context.cjs');
require('../../../../../context/FooterAnimationContext/index.cjs');
var Typography = require('../../../../../components/Typography/Typography.cjs');
require('../../../../../context/SocialRedirectContext/SocialRedirectContext.cjs');
require('../../../../../context/WalletGroupContext/WalletGroupContext.cjs');
require('../../../components/DynamicWidgetHeader/DynamicWidgetHeader.cjs');
require('react-focus-lock');
require('../../../../../context/LoadingContext/LoadingContext.cjs');
var getPasskeyProviderFromUserAgent = require('../../../helpers/getPasskeyProviderFromUserAgent.cjs');
var getTimeSince = require('../../../helpers/getTimeSince.cjs');
require('../../../context/DynamicWidgetContext.cjs');
require('../../../../../components/UserProfile/parts/UserProfileField/components/VerifiedEmailIcon/VerifiedEmailIcon.cjs');
require('formik');
require('../../../../../components/IconButton/IconButton.cjs');
require('../../../../../components/Alert/Alert.cjs');
require('../../../../../components/PasskeyCreatedSuccessBanner/PasskeyCreatedSuccessBanner.cjs');
var PasskeyContext = require('../../../../../context/PasskeyContext/PasskeyContext.cjs');
require('@dynamic-labs/viem-utils');
require('../../../../../views/TransactionConfirmationView/helpers/transactionErrorMessage.cjs');
require('../../../../../utils/hooks/useTransactionWithGasPrice/useTransactionWithGasPrice.cjs');
require('../../../../../components/InlineWidget/InlineWidget.cjs');
require('../../../../../components/IsBrowser/IsBrowser.cjs');
require('../../../../../components/MenuList/Dropdown/Dropdown.cjs');
require('../../../../../components/Popper/Popper/Popper.cjs');
require('../../../../../components/Popper/PopperContext/PopperContext.cjs');
require('qrcode');
require('../../../../../views/WalletList/WalletList.cjs');
require('@hcaptcha/react-hcaptcha');
require('../../../../DynamicBridgeWidget/views/WalletsView/components/SecondaryWallets/SecondaryWallets.cjs');

const mapPasskeyProviderToIcon = {
    android: jsxRuntime.jsx(iconic.AndroidIcon, {}),
    brave: jsxRuntime.jsx(iconic.BraveIcon, {}),
    chrome: jsxRuntime.jsx(iconic.ChromeIcon, {}),
    edge: jsxRuntime.jsx(iconic.EdgeIcon, {}),
    firefox: jsxRuntime.jsx(iconic.FirefoxIcon, {}),
    iPhone: jsxRuntime.jsx(iconic.AppleIcon, {}),
    opera: jsxRuntime.jsx(iconic.OperaIcon, {}),
    safari: jsxRuntime.jsx(iconic.SafariIcon, {}),
};
const PasskeyCard = ({ passkey }) => {
    var _a;
    const { t } = reactI18next.useTranslation();
    const { setShowAuthFlow } = useInternalDynamicContext.useInternalDynamicContext();
    const { setView } = ViewContext.useViewContext();
    const { setPasskey, setPasskeyIcon } = PasskeyContext.usePasskeyContext();
    const menuOptions = [
        {
            Icon: null,
            callback: () => {
                setShowAuthFlow(true);
                setPasskey(passkey);
                setPasskeyIcon(() => { var _a; return (_a = PasskeyProviderIcon()) !== null && _a !== void 0 ? _a : undefined; });
                setView('rename-passkey');
            },
            text: 'Rename',
        },
    ];
    const [showTooltip, setShowTooltip] = React.useState(false);
    const { value: timeSinceValue, unit: timeSinceUnit } = getTimeSince.getTimeSince(passkey.createdAt);
    const PasskeyProviderIcon = () => {
        if (!passkey.userAgent)
            return null;
        const provider = getPasskeyProviderFromUserAgent.getPasskeyProviderFromUserAgent(passkey.userAgent);
        if (!provider)
            return null;
        return mapPasskeyProviderToIcon[provider];
    };
    const getPasskeyProviderName = () => {
        if (!passkey.userAgent)
            return null;
        const provider = getPasskeyProviderFromUserAgent.getPasskeyProviderFromUserAgent(passkey.userAgent);
        if (!provider)
            return null;
        return t(`dyn_manage_passkeys.passkey_providers.${provider}`);
    };
    const handleShowTooltip = () => {
        setShowTooltip(true);
        setTimeout(() => {
            setShowTooltip(false);
        }, 2000);
    };
    return (jsxRuntime.jsxs("div", { className: 'passkey-card', children: [jsxRuntime.jsx("div", { className: 'passkey-card__info', children: jsxRuntime.jsxs("div", { className: 'passkey-card__info__text', children: [jsxRuntime.jsxs(Typography.Typography, { "data-testid": 'passkey-alias', variant: 'body_normal', color: 'primary', title: passkey.alias, onClick: handleShowTooltip, children: [jsxRuntime.jsxs("div", { className: 'passkey-card__body', children: [jsxRuntime.jsxs("div", { className: 'passkey-card__left', children: [jsxRuntime.jsx("div", { className: 'passkey-card__info__icon', "data-testid": 'provider-icon', children: jsxRuntime.jsx(PasskeyProviderIcon, {}) }), ((_a = passkey.alias) === null || _a === void 0 ? void 0 : _a.includes('@'))
                                                    ? shortenEmail.shortenEmail(passkey.alias, {
                                                        breakpoint: 12,
                                                        cutEnd: 4,
                                                        cutStart: 4,
                                                    })
                                                    : passkey.alias] }), jsxRuntime.jsx(DotsMenu.DotsMenu, { options: menuOptions })] }), jsxRuntime.jsx("span", { className: classNames.classNames('passkey-card__info__tooltip', {
                                        'passkey-card__info__tooltip--visible': showTooltip,
                                    }), children: passkey.alias })] }), getPasskeyProviderName() && (jsxRuntime.jsxs(Typography.Typography, { variant: 'body_small', color: 'secondary', children: [t('dyn_manage_passkeys.passkey_from'), " ", getPasskeyProviderName()] }))] }) }), jsxRuntime.jsx("div", { className: 'passkey-card__date', children: jsxRuntime.jsxs(Typography.Typography, { variant: 'body_small', color: 'secondary', copykey: 'dyn_time_since.created', children: [t('dyn_time_since.created'), ' ', `${timeSinceValue} ${timeSinceUnit} ${t('dyn_time_since.ago')}`] }) })] }));
};

exports.PasskeyCard = PasskeyCard;
