// @zerodev/ecdsa-validator/_cjs/toECDSAValidatorPlugin.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.signerToEcdsaValidator = void 0;
const sdk_1 = require("@zerodev/sdk");
const permissionless_1 = require("permissionless");
const accounts_1 = require("permissionless/accounts");
const accounts_2 = require("viem/accounts");
const actions_1 = require("viem/actions");
const actions_2 = require("viem/actions");
const index_js_1 = require("./index.js");
async function signerToEcdsaValidator(client, { signer, entryPoint = sdk_1.KERNEL_ADDRESSES.ENTRYPOINT_V0_6, validatorAddress = index_js_1.ECDSA_VALIDATOR_ADDRESS }) {
    const viemSigner = {
        ...signer,
        signTransaction: (_, __) => {
            throw new accounts_1.SignTransactionNotSupportedBySmartAccount();
        }
    };
    const chainId = await (0, actions_2.getChainId)(client);
    const account = (0, accounts_2.toAccount)({
        address: viemSigner.address,
        async signMessage({ message }) {
            return (0, actions_1.signMessage)(client, { account: viemSigner, message });
        },
        async signTransaction(_, __) {
            throw new accounts_1.SignTransactionNotSupportedBySmartAccount();
        },
        async signTypedData(typedData) {
            return (0, actions_1.signTypedData)(client, {
                account: viemSigner,
                ...typedData
            });
        }
    });
    return {
        ...account,
        address: validatorAddress,
        source: "ECDSAValidator",
        async getEnableData() {
            return viemSigner.address;
        },
        async getNonceKey() {
            return 0n;
        },
        async signUserOperation(userOperation) {
            const hash = (0, permissionless_1.getUserOperationHash)({
                userOperation: {
                    ...userOperation,
                    signature: "0x"
                },
                entryPoint: entryPoint,
                chainId: chainId
            });
            const signature = await (0, actions_1.signMessage)(client, {
                account: viemSigner,
                message: { raw: hash }
            });
            return signature;
        },
        async getDummySignature() {
            return "0xfffffffffffffffffffffffffffffff0000000000000000000000000000000007aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa1c";
        },
        async isEnabled(_kernelAccountAddress, _selector) {
            return false;
        }
    };
}
exports.signerToEcdsaValidator = signerToEcdsaValidator;
