// @dynamic-labs/ethereum/src/index.cjs
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

require('./polyfills.cjs');
var turnkey = require('@dynamic-labs/turnkey');
var PhantomEvm = require('./injected/PhantomEvm.cjs');
var ExodusEvm = require('./injected/ExodusEvm.cjs');
var fetchInjectedWalletConnectors = require('./injected/fetchInjectedWalletConnectors.cjs');
require('./walletConnect/walletConnectV2.cjs');
require('../_virtual/_tslib.cjs');
var EthWalletConnector = require('./EthWalletConnector.cjs');
var fetchWalletConnectWallets = require('./walletConnect/fetchWalletConnectWallets.cjs');
var coinbase = require('./coinbase/coinbase.cjs');
var ethProviderHelper = require('./ethProviderHelper.cjs');
var constants = require('./constants.cjs');

const EthereumWalletConnectors = (props) => [
    ...fetchInjectedWalletConnectors.injectedWalletOverrides,
    ...fetchInjectedWalletConnectors.fetchInjectedWalletConnector(props),
    ...fetchWalletConnectWallets.fetchWalletConnectWallets(props),
    ...turnkey.TurnkeyEVMWalletConnectors(props),
    coinbase.Coinbase,
    fetchWalletConnectWallets.getWalletConnectConnector(props),
];

exports.PhantomEvm = PhantomEvm.PhantomEvm;
exports.ExodusEvm = ExodusEvm.ExodusEvm;
exports.fetchInjectedWalletConnector = fetchInjectedWalletConnectors.fetchInjectedWalletConnector;
exports.injectedWalletOverrides = fetchInjectedWalletConnectors.injectedWalletOverrides;
exports.EthWalletConnector = EthWalletConnector.EthWalletConnector;
exports.EthProviderHelper = ethProviderHelper.EthProviderHelper;
exports.INFURA_ID = constants.INFURA_ID;
exports.EthereumWalletConnectors = EthereumWalletConnectors;
