// @dynamic-labs/ethereum/src/index.js
import './polyfills.js';
import { TurnkeyEVMWalletConnectors } from '@dynamic-labs/turnkey';
export { PhantomEvm } from './injected/PhantomEvm.js';
export { ExodusEvm } from './injected/ExodusEvm.js';
import { injectedWalletOverrides, fetchInjectedWalletConnector } from './injected/fetchInjectedWalletConnectors.js';
export { fetchInjectedWalletConnector, injectedWalletOverrides } from './injected/fetchInjectedWalletConnectors.js';
import './walletConnect/walletConnectV2.js';
import '../_virtual/_tslib.js';
export { EthWalletConnector } from './EthWalletConnector.js';
import { fetchWalletConnectWallets, getWalletConnectConnector } from './walletConnect/fetchWalletConnectWallets.js';
import { Coinbase } from './coinbase/coinbase.js';
export { EthProviderHelper } from './ethProviderHelper.js';
export { INFURA_ID } from './constants.js';

const EthereumWalletConnectors = (props) => [
    ...injectedWalletOverrides,
    ...fetchInjectedWalletConnector(props),
    ...fetchWalletConnectWallets(props),
    ...TurnkeyEVMWalletConnectors(props),
    Coinbase,
    getWalletConnectConnector(props),
];

export { EthereumWalletConnectors };
