// @alchemy/aa-core/dist/esm/client/create-client.js
import { createPublicClient, http, } from "viem";
import { VERSION } from "../version.js";
export const createPublicErc4337FromClient = (client) => {
    const clientAdapter = client;
    return {
        ...clientAdapter,
        estimateUserOperationGas(request, entryPoint) {
            return clientAdapter.request({
                method: "eth_estimateUserOperationGas",
                params: [request, entryPoint],
            });
        },
        sendUserOperation(request, entryPoint) {
            return clientAdapter.request({
                method: "eth_sendUserOperation",
                params: [request, entryPoint],
            });
        },
        getUserOperationByHash(hash) {
            return clientAdapter.request({
                method: "eth_getUserOperationByHash",
                params: [hash],
            });
        },
        getUserOperationReceipt(hash) {
            return clientAdapter.request({
                method: "eth_getUserOperationReceipt",
                params: [hash],
            });
        },
        getSupportedEntryPoints() {
            return clientAdapter.request({
                method: "eth_supportedEntryPoints",
                params: [],
            });
        },
        getMaxPriorityFeePerGas() {
            return clientAdapter.request({
                method: "eth_maxPriorityFeePerGas",
                params: [],
            });
        },
        async getFeeData() {
            const block = await clientAdapter.getBlock({
                blockTag: "latest",
            });
            if (block && block.baseFeePerGas) {
                const maxPriorityFeePerGas = BigInt(1500000000);
                return {
                    maxPriorityFeePerGas,
                    maxFeePerGas: block.baseFeePerGas * BigInt(2) + maxPriorityFeePerGas,
                };
            }
            return {
                maxFeePerGas: 0,
                maxPriorityFeePerGas: 0,
            };
        },
        async getContractCode(address) {
            const code = await clientAdapter.getBytecode({
                address: address,
            });
            return code ?? "0x";
        },
    };
};
export const createPublicErc4337Client = ({ chain, rpcUrl, fetchOptions, }) => {
    const client = createPublicErc4337FromClient(createPublicClient({
        chain,
        transport: http(rpcUrl, {
            fetchOptions: {
                ...fetchOptions,
                headers: {
                    ...fetchOptions?.headers,
                    "Alchemy-AA-Sdk-Version": VERSION,
                },
            },
        }),
    }));
    return client;
};
