// @zerodev/sdk/dist/esm/kernel-zerodev/validator-provider/ecdsa-provider.js
import { ValidatorProvider, } from "./base.js";
import { ECDSAValidator, } from "../validator/ecdsa-validator.js";
import { getChainId } from "../api/index.js";
import { polygonMumbai } from "viem/chains";
import { getChain } from "../utils.js";
export class ECDSAProvider extends ValidatorProvider {
    constructor(params) {
        const chain = typeof params.opts?.providerConfig?.chain === "number"
            ? getChain(params.opts.providerConfig.chain)
            : params.opts?.providerConfig?.chain ?? polygonMumbai;
        const validator = new ECDSAValidator({
            projectId: params.projectId,
            owner: params.owner,
            chain,
            rpcUrl: params.opts?.providerConfig?.rpcUrl,
            ...params.opts?.validatorConfig,
        });
        super({
            ...params,
            opts: {
                ...params.opts,
                providerConfig: { ...params.opts?.providerConfig, chain },
            },
        }, validator);
        Object.defineProperty(this, "changeOwner", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: this.sendEnableUserOperation
        });
        Object.defineProperty(this, "deleteOwner", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: this.sendDisableUserOperation
        });
    }
    static async init(params) {
        const chainId = await getChainId(params.projectId);
        if (!chainId) {
            throw new Error("ChainId not found");
        }
        const chain = getChain(chainId);
        const instance = new ECDSAProvider({
            ...params,
            opts: {
                ...params.opts,
                providerConfig: {
                    chain,
                    ...params.opts?.providerConfig,
                },
            },
        });
        return instance;
    }
}
