// @zerodev/sdk/dist/esm/kernel-zerodev/validator-provider/kill-switch-provider.js
import { ValidatorProvider, } from "./base.js";
import { KillSwitchValidator, } from "../validator/kill-switch-validator.js";
import { getChainId } from "../api/index.js";
import { polygonMumbai } from "viem/chains";
import { KILL_SWITCH_VALIDATOR_ADDRESS } from "../constants.js";
import { getChain } from "../utils.js";
export class KillSwitchProvider extends ValidatorProvider {
    constructor(params) {
        const chain = typeof params.opts?.providerConfig?.chain === "number"
            ? getChain(params.opts.providerConfig.chain)
            : params.opts?.providerConfig?.chain ?? polygonMumbai;
        const validator = new KillSwitchValidator({
            projectId: params.projectId,
            guardian: params.guardian,
            delaySeconds: params.delaySeconds,
            chain,
            rpcUrl: params.opts?.providerConfig?.rpcUrl,
            validatorAddress: params.opts?.validatorConfig?.validatorAddress ??
                KILL_SWITCH_VALIDATOR_ADDRESS,
            ...params.opts?.validatorConfig,
        });
        super({
            ...params,
            opts: {
                ...params.opts,
                providerConfig: { ...params.opts?.providerConfig, chain },
            },
        }, validator);
        Object.defineProperty(this, "changeGuardian", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: this.sendEnableUserOperation
        });
        Object.defineProperty(this, "deleteKillSwitchData", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: this.sendDisableUserOperation
        });
    }
    static async init(params) {
        const chainId = await getChainId(params.projectId);
        if (!chainId) {
            throw new Error("ChainId not found");
        }
        const chain = getChain(chainId);
        const instance = new KillSwitchProvider({
            ...params,
            opts: {
                ...params.opts,
                providerConfig: {
                    chain,
                    ...params.opts?.providerConfig,
                },
            },
        });
        return instance;
    }
}
