// @dynamic-labs/sdk-react-core/src/lib/shared/assets/alt-embedded-wallet.cjs
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var React = require('react');

function _interopNamespace(e) {
  if (e && e.__esModule) return e;
  var n = Object.create(null);
  if (e) {
    Object.keys(e).forEach(function (k) {
      if (k !== 'default') {
        var d = Object.getOwnPropertyDescriptor(e, k);
        Object.defineProperty(n, k, d.get ? d : {
          enumerable: true,
          get: function () { return e[k]; }
        });
      }
    });
  }
  n["default"] = e;
  return Object.freeze(n);
}

var React__namespace = /*#__PURE__*/_interopNamespace(React);

var _path;
var _excluded = ["title", "titleId"];
function _extends() { _extends = Object.assign ? Object.assign.bind() : function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; return _extends.apply(this, arguments); }
function _objectWithoutProperties(source, excluded) { if (source == null) return {}; var target = _objectWithoutPropertiesLoose(source, excluded); var key, i; if (Object.getOwnPropertySymbols) { var sourceSymbolKeys = Object.getOwnPropertySymbols(source); for (i = 0; i < sourceSymbolKeys.length; i++) { key = sourceSymbolKeys[i]; if (excluded.indexOf(key) >= 0) continue; if (!Object.prototype.propertyIsEnumerable.call(source, key)) continue; target[key] = source[key]; } } return target; }
function _objectWithoutPropertiesLoose(source, excluded) { if (source == null) return {}; var target = {}; var sourceKeys = Object.keys(source); var key, i; for (i = 0; i < sourceKeys.length; i++) { key = sourceKeys[i]; if (excluded.indexOf(key) >= 0) continue; target[key] = source[key]; } return target; }
var SvgAltEmbeddedWallet = function SvgAltEmbeddedWallet(_ref) {
  var title = _ref.title,
    titleId = _ref.titleId,
    props = _objectWithoutProperties(_ref, _excluded);
  return /*#__PURE__*/React__namespace.createElement("svg", _extends({
    width: 28,
    height: 28,
    viewBox: "0 0 28 28",
    fill: "none",
    xmlns: "http://www.w3.org/2000/svg",
    "aria-labelledby": titleId
  }, props), title ? /*#__PURE__*/React__namespace.createElement("title", {
    id: titleId
  }, title) : null, _path || (_path = /*#__PURE__*/React__namespace.createElement("path", {
    fillRule: "evenodd",
    clipRule: "evenodd",
    d: "M19.834 4.667a2.333 2.333 0 0 0-2.333-2.334H7a4.668 4.668 0 0 0-4.592 3.83 4.654 4.654 0 0 1 3.425-1.496h14ZM21 7a4.667 4.667 0 0 1 4.667 4.667V21A4.667 4.667 0 0 1 21 25.667H7A4.667 4.667 0 0 1 2.333 21v-9.333A4.667 4.667 0 0 1 7 7h14Zm-5.281 3.35a.506.506 0 0 0-.47-.35.506.506 0 0 0-.47.35l-.593 1.671c-.152.426-.2.549-.264.648-.065.1-.146.186-.237.257-.092.07-.205.122-.598.286l-1.543.643a.546.546 0 0 0-.323.509c0 .225.128.428.323.509l1.543.643c.393.163.506.215.598.285.091.071.172.158.237.257.065.1.112.222.264.648l.593 1.672c.075.21.262.35.47.35a.506.506 0 0 0 .47-.35l.593-1.672c.152-.426.2-.549.264-.648.065-.1.145-.186.237-.257.092-.07.205-.122.598-.286l1.543-.642a.546.546 0 0 0 .323-.51.546.546 0 0 0-.323-.508l-1.543-.643c-.393-.164-.506-.216-.598-.286a1.054 1.054 0 0 1-.237-.257c-.065-.1-.112-.222-.264-.648l-.593-1.671Zm-4.3 5.951a.501.501 0 0 0-.45-.301.5.5 0 0 0-.45.301l-.395.856c-.142.308-.186.397-.238.471-.053.074-.115.14-.183.198-.068.057-.15.104-.435.258l-.79.428A.552.552 0 0 0 8.2 19c0 .206.108.395.278.488l.79.428c.285.154.367.2.435.258.068.057.13.123.183.198.052.074.096.163.238.47l.395.856a.5.5 0 0 0 .45.302c.191 0 .365-.117.45-.302l.396-.855c.142-.308.185-.397.238-.471.053-.075.114-.141.183-.198.068-.057.15-.104.434-.258l.79-.428a.552.552 0 0 0 .278-.488.552.552 0 0 0-.278-.488l-.79-.428c-.284-.154-.366-.2-.434-.258a1.05 1.05 0 0 1-.183-.198c-.053-.074-.096-.163-.238-.47l-.395-.857Z",
    fill: "#383C48",
    fillOpacity: 0.4
  })));
};

exports.ReactComponent = SvgAltEmbeddedWallet;
