// @dynamic-labs/sdk-react-core/src/lib/utils/hooks/useConnectWallet/useConnectWallet.cjs
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var _tslib = require('../../../../../_virtual/_tslib.cjs');
var React = require('react');
var utils = require('@dynamic-labs/utils');
var walletConnectorCore = require('@dynamic-labs/wallet-connector-core');
var UserWalletsContext = require('../../../context/UserWalletsContext/UserWalletsContext.cjs');
var api = require('../../../data/api.cjs');
require('@dynamic-labs/sdk-api');
require('../../../shared/logger.cjs');
require('@dynamic-labs/iconic');
require('react/jsx-runtime');
require('../../../context/ViewContext/ViewContext.cjs');
require('@dynamic-labs/wallet-book');
var localStorage = require('../../constants/localStorage.cjs');
require('../../constants/colors.cjs');
require('../../constants/values.cjs');
require('../../../shared/utils/classes/storage/localStorage.cjs');
require('viem');
var useLocalStorage = require('../../../shared/utils/hooks/useLocalStorage/useLocalStorage.cjs');
require('../../../shared/consts/index.cjs');
require('@dynamic-labs/multi-wallet');
var getWalletIdentifier = require('../../functions/getWalletIdentifier/getWalletIdentifier.cjs');
var shouldManuallyReconnectOnRefresh = require('../../functions/shouldManuallyReconnectOnRefresh/shouldManuallyReconnectOnRefresh.cjs');
require('viem/chains');
var getWalletProvider = require('../../functions/getWalletProvider/getWalletProvider.cjs');
var isConnectOnly = require('../authenticationHooks/helpers/isConnectOnly.cjs');
var dynamicEvents = require('../../../events/dynamicEvents.cjs');
var updateUserWalletsFromConnectedWallets = require('./updateUserWalletsFromConnectedWallets/updateUserWalletsFromConnectedWallets.cjs');

const useConnectWallet = ({ authMode, clearPrimaryWalletId, enableVisitTrackingOnConnectOnly, environmentId, primaryWalletId, setPrimaryWalletId, walletConnectorOptions, onDisconnect, onBeforeConnectSuccessConfirmation, setShowAuthFlow, isBridgeFlow, setIsVerificationInProgress, user, }) => {
    const [connectedWalletsInfo, setConnectedWalletsInfo, removeConnectedWalletsInfo,] = useLocalStorage.useLocalStorage(localStorage.CONNECTED_WALLETS_INFO, []);
    const [connectedWallets, _setConnectedWallets] = React.useState([]);
    /**
     * When a new wallet is connected with the onConnect callback enabled,
     * this new wallet's id will hang in this array until we attest that the corresponding wallet object
     * has been successfully built — at which point we raise the callback and pop the id from this array.
     * Should be removed with QNTM-1082 refactor
     */
    const onConnectWaitList = React.useRef(new Set());
    const { setUserWallets } = UserWalletsContext.useInternalUserWallets();
    /** This wrapper is to ensure userWallets is always properly updated alongside connectedWallets */
    const setConnectedWallets = (newWallets) => {
        _setConnectedWallets(newWallets);
        setUserWallets((userWallets) => updateUserWalletsFromConnectedWallets.updateUserWalletsFromConnectedWallets(userWallets, newWallets));
    };
    const disconnectWallet = React.useCallback((walletId) => _tslib.__awaiter(void 0, void 0, void 0, function* () {
        const walletToDisconnect = connectedWallets.find((wallet) => wallet.id === walletId);
        yield (walletToDisconnect === null || walletToDisconnect === void 0 ? void 0 : walletToDisconnect.connector.endSession());
        const updatedConnectedWalletsInfo = connectedWalletsInfo.filter((wallet) => wallet.id !== walletId);
        setConnectedWalletsInfo(updatedConnectedWalletsInfo);
        onDisconnect(walletToDisconnect);
        if (walletId !== primaryWalletId) {
            return;
        }
        const hasConnectedWallet = updatedConnectedWalletsInfo.length > 0;
        if (isBridgeFlow || !hasConnectedWallet) {
            clearPrimaryWalletId();
        }
        else {
            setPrimaryWalletId(updatedConnectedWalletsInfo[0].id);
        }
    }), [
        clearPrimaryWalletId,
        connectedWallets,
        connectedWalletsInfo,
        isBridgeFlow,
        onDisconnect,
        primaryWalletId,
        setConnectedWalletsInfo,
        setPrimaryWalletId,
    ]);
    // The function to update and define the connectedWallets list.
    // It uses walletConnector to get the current wallet address and network
    // The address and network in the runtime are updated by the event listeners
    // The event listeners are updating `connectedWallets` list.
    // It should be called on the first render and when connectedWalletsInfo updates
    const updateConnectedWalletsList = React.useCallback(() => _tslib.__awaiter(void 0, void 0, void 0, function* () {
        const walletConnectors = walletConnectorOptions.map((wallet) => wallet.walletConnector);
        if (!walletConnectors.length) {
            return;
        }
        const updatedConnectedWallets = (yield Promise.all(connectedWalletsInfo.map((storedConnectedWalletInfo) => _tslib.__awaiter(void 0, void 0, void 0, function* () {
            const walletConnector = walletConnectorCore.getWalletConnectorByKey(walletConnectors, storedConnectedWalletInfo.walletConnectorKey);
            if (!walletConnector) {
                walletConnectorCore.logger.error('Could not find walletConnector: ' +
                    storedConnectedWalletInfo.walletConnectorKey);
                yield disconnectWallet(storedConnectedWalletInfo.id);
                return null;
            }
            const [walletAddress] = yield walletConnector.getConnectedAccounts();
            if (!walletAddress) {
                yield disconnectWallet(storedConnectedWalletInfo.id);
                return null;
            }
            const [walletNetwork, additionalAddresses] = yield Promise.all([
                walletConnector.getNetwork(),
                walletConnector.getAdditionalAddresses(walletAddress),
            ]);
            const walletChain = walletConnector.connectedChain;
            const walletObject = {
                additionalAddresses,
                address: walletAddress || '',
                authenticated: false,
                chain: walletChain || '',
                connected: true,
                connector: walletConnector,
                id: storedConnectedWalletInfo.id,
                network: walletNetwork,
            };
            return walletObject;
        })))).filter((wallet) => Boolean(wallet));
        setConnectedWallets(updatedConnectedWallets);
        // Raise any onConnect callbacks in wait list
        for (const walletId of [...onConnectWaitList.current]) {
            const wallet = updatedConnectedWallets.find((wallet) => getWalletIdentifier.getWalletIdentifier(wallet) === walletId);
            if (!wallet)
                continue;
            onConnectWaitList.current.delete(walletId);
            dynamicEvents.dynamicEvents.emit('walletConnected', wallet);
            setIsVerificationInProgress(false);
        }
    }), [
        connectedWalletsInfo,
        setIsVerificationInProgress,
        disconnectWallet,
        walletConnectorOptions,
    ]);
    // Generate the connectedWallets list when connectedWalletsInfo (localStorage) or memoized wallet connector updates
    React.useEffect(() => {
        updateConnectedWalletsList();
        // eslint-disable-next-line react-hooks/exhaustive-deps
    }, [connectedWalletsInfo, walletConnectorOptions]);
    const getNextIdAvailable = React.useCallback((currentWallets) => {
        const existingIds = currentWallets.map(({ id }) => id);
        const sortedIds = existingIds
            .map((id) => Number(id.replace('connect-wallet-', '')))
            .sort((a, b) => a - b);
        for (let i = 0; i < sortedIds.length; i++) {
            if (sortedIds[i] !== i) {
                return i;
            }
        }
        return sortedIds.length;
    }, []);
    const updateConnectedWalletById = React.useCallback((walletId, walletObject) => {
        const clonedConnectedWalletsList = [...connectedWallets];
        const connectedWalletIndex = clonedConnectedWalletsList.findIndex((connectedWallet) => connectedWallet.id === walletId);
        if (connectedWalletIndex > -1) {
            if (walletObject.address && walletObject.network) {
                clonedConnectedWalletsList[connectedWalletIndex] = Object.assign(Object.assign({}, clonedConnectedWalletsList[connectedWalletIndex]), { address: walletObject.address, network: walletObject.network });
            }
            else if (walletObject.address) {
                clonedConnectedWalletsList[connectedWalletIndex] = Object.assign(Object.assign({}, clonedConnectedWalletsList[connectedWalletIndex]), { address: walletObject.address });
            }
            else {
                clonedConnectedWalletsList[connectedWalletIndex] = Object.assign(Object.assign({}, clonedConnectedWalletsList[connectedWalletIndex]), { network: walletObject.network });
            }
            setConnectedWallets(clonedConnectedWalletsList);
        }
    }, [connectedWallets]);
    // Keeps connected wallet data inside localStorage
    const applyConnectedWalletToStore = ({ walletId, walletConnectorKey, walletChain, provider, }) => {
        const updatedConnectedWalletsInfo = [...connectedWalletsInfo];
        updatedConnectedWalletsInfo.push({
            id: walletId,
            provider,
            walletChain,
            walletConnectorKey,
        });
        setConnectedWalletsInfo(updatedConnectedWalletsInfo);
    };
    // Updates connected wallet (network and address) and verifies connected wallet connection
    const refreshConnectedWallet = (walletId, walletConnector) => _tslib.__awaiter(void 0, void 0, void 0, function* () {
        if (shouldManuallyReconnectOnRefresh.shouldManuallyReconnectOnRefresh(walletConnector)) {
            yield walletConnector.connect();
        }
        const currentWalletConnectorAddress = yield walletConnector.fetchPublicAddress();
        if (currentWalletConnectorAddress && onBeforeConnectSuccessConfirmation) {
            const shouldProceedWithConnection = yield onBeforeConnectSuccessConfirmation({
                address: currentWalletConnectorAddress,
                chain: walletConnector.connectedChain,
                connector: walletConnector,
            });
            if (!shouldProceedWithConnection) {
                walletConnectorCore.logger.info('Connection was not established because onBeforeConnectSuccessConfirmation returned false');
                disconnectWallet(walletId);
                setShowAuthFlow(false, { emitCancelAuth: true });
                return;
            }
        }
        const currentWalletConnectorNetwork = yield walletConnector.getNetwork();
        updateConnectedWalletById(walletId, {
            address: currentWalletConnectorAddress,
            network: currentWalletConnectorNetwork,
        });
    });
    const shouldSkipCreatingVisit = authMode === 'connect-only' && !enableVisitTrackingOnConnectOnly;
    const connectWallet = (walletConnector, fetchPublicAddressOpts, options) => _tslib.__awaiter(void 0, void 0, void 0, function* () {
        const { applyOnBeforeConnectSuccessConfirmation = true, raiseOnConnect = false, } = options !== null && options !== void 0 ? options : {};
        const walletAddress = yield walletConnector.fetchPublicAddress(fetchPublicAddressOpts);
        if (!walletAddress) {
            throw new utils.MissingPublicAddressError();
        }
        if (onBeforeConnectSuccessConfirmation &&
            applyOnBeforeConnectSuccessConfirmation) {
            const shouldProceedWithConnection = yield onBeforeConnectSuccessConfirmation({
                address: walletAddress,
                chain: walletConnector.connectedChain,
                connector: walletConnector,
            });
            if (!shouldProceedWithConnection) {
                walletConnectorCore.logger.info('Connection was not established because onBeforeConnectSuccessConfirmation returned false');
                setShowAuthFlow(false);
                return;
            }
        }
        const walletId = `connect-wallet-${getNextIdAvailable(connectedWalletsInfo)}`;
        // We are already using the new syntax for wallet ids for this.
        // Eventually the connect-wallet-X syntax will be dropped
        if (raiseOnConnect && isConnectOnly.isConnectOnly(user, authMode))
            onConnectWaitList.current.add(getWalletIdentifier.getWalletIdentifier({
                address: walletAddress,
                connector: walletConnector,
            }));
        if (!primaryWalletId) {
            setPrimaryWalletId(walletId);
        }
        const walletChain = walletConnector.connectedChain;
        const isWalletStored = connectedWalletsInfo.some(({ walletConnectorKey: storedWalletConnectorKey }) => storedWalletConnectorKey === walletConnector.key);
        if (isWalletStored) {
            yield updateConnectedWalletsList();
        }
        else {
            applyConnectedWalletToStore({
                provider: getWalletProvider.getWalletProvider(walletConnector),
                walletAddress: walletAddress,
                walletChain,
                walletConnectorKey: walletConnector.key,
                walletId,
            });
            if (!shouldSkipCreatingVisit) {
                // send information to backend to kick off background jobs
                // so verify/sign on the next step could go by more quickly
                // this is async work, but DO NOT AWAIT
                api.createVisit({
                    authMode,
                    chain: walletChain || '',
                    environmentId,
                    publicWalletAddress: walletAddress,
                    walletName: walletConnector.key,
                    walletProvider: getWalletProvider.getWalletProvider(walletConnector),
                });
            }
        }
        return { address: walletAddress, id: walletId };
    });
    const getConnectedWalletById = React.useCallback((walletId) => connectedWallets.find((wallet) => wallet.id === walletId), [connectedWallets]);
    return {
        connectWallet,
        connectedWallets,
        connectedWalletsInfo,
        disconnectWallet,
        getConnectedWalletById,
        refreshConnectedWallet,
        removeConnectedWalletsInfo,
    };
};

exports.useConnectWallet = useConnectWallet;
