// @dynamic-labs/sdk-react-core/src/lib/utils/hooks/useConnectWallet/useConnectWallet.js
import { __awaiter } from '../../../../../_virtual/_tslib.js';
import { useState, useRef, useCallback, useEffect } from 'react';
import { MissingPublicAddressError } from '@dynamic-labs/utils';
import { getWalletConnectorByKey, logger } from '@dynamic-labs/wallet-connector-core';
import { useInternalUserWallets } from '../../../context/UserWalletsContext/UserWalletsContext.js';
import { createVisit } from '../../../data/api.js';
import '@dynamic-labs/sdk-api';
import '../../../shared/logger.js';
import '@dynamic-labs/iconic';
import 'react/jsx-runtime';
import '../../../context/ViewContext/ViewContext.js';
import '@dynamic-labs/wallet-book';
import { CONNECTED_WALLETS_INFO } from '../../constants/localStorage.js';
import '../../constants/colors.js';
import '../../constants/values.js';
import '../../../shared/utils/classes/storage/localStorage.js';
import 'viem';
import { useLocalStorage } from '../../../shared/utils/hooks/useLocalStorage/useLocalStorage.js';
import '../../../shared/consts/index.js';
import '@dynamic-labs/multi-wallet';
import { getWalletIdentifier } from '../../functions/getWalletIdentifier/getWalletIdentifier.js';
import { shouldManuallyReconnectOnRefresh } from '../../functions/shouldManuallyReconnectOnRefresh/shouldManuallyReconnectOnRefresh.js';
import 'viem/chains';
import { getWalletProvider } from '../../functions/getWalletProvider/getWalletProvider.js';
import { isConnectOnly } from '../authenticationHooks/helpers/isConnectOnly.js';
import { dynamicEvents } from '../../../events/dynamicEvents.js';
import { updateUserWalletsFromConnectedWallets } from './updateUserWalletsFromConnectedWallets/updateUserWalletsFromConnectedWallets.js';

const useConnectWallet = ({ authMode, clearPrimaryWalletId, enableVisitTrackingOnConnectOnly, environmentId, primaryWalletId, setPrimaryWalletId, walletConnectorOptions, onDisconnect, onBeforeConnectSuccessConfirmation, setShowAuthFlow, isBridgeFlow, setIsVerificationInProgress, user, }) => {
    const [connectedWalletsInfo, setConnectedWalletsInfo, removeConnectedWalletsInfo,] = useLocalStorage(CONNECTED_WALLETS_INFO, []);
    const [connectedWallets, _setConnectedWallets] = useState([]);
    /**
     * When a new wallet is connected with the onConnect callback enabled,
     * this new wallet's id will hang in this array until we attest that the corresponding wallet object
     * has been successfully built — at which point we raise the callback and pop the id from this array.
     * Should be removed with QNTM-1082 refactor
     */
    const onConnectWaitList = useRef(new Set());
    const { setUserWallets } = useInternalUserWallets();
    /** This wrapper is to ensure userWallets is always properly updated alongside connectedWallets */
    const setConnectedWallets = (newWallets) => {
        _setConnectedWallets(newWallets);
        setUserWallets((userWallets) => updateUserWalletsFromConnectedWallets(userWallets, newWallets));
    };
    const disconnectWallet = useCallback((walletId) => __awaiter(void 0, void 0, void 0, function* () {
        const walletToDisconnect = connectedWallets.find((wallet) => wallet.id === walletId);
        yield (walletToDisconnect === null || walletToDisconnect === void 0 ? void 0 : walletToDisconnect.connector.endSession());
        const updatedConnectedWalletsInfo = connectedWalletsInfo.filter((wallet) => wallet.id !== walletId);
        setConnectedWalletsInfo(updatedConnectedWalletsInfo);
        onDisconnect(walletToDisconnect);
        if (walletId !== primaryWalletId) {
            return;
        }
        const hasConnectedWallet = updatedConnectedWalletsInfo.length > 0;
        if (isBridgeFlow || !hasConnectedWallet) {
            clearPrimaryWalletId();
        }
        else {
            setPrimaryWalletId(updatedConnectedWalletsInfo[0].id);
        }
    }), [
        clearPrimaryWalletId,
        connectedWallets,
        connectedWalletsInfo,
        isBridgeFlow,
        onDisconnect,
        primaryWalletId,
        setConnectedWalletsInfo,
        setPrimaryWalletId,
    ]);
    // The function to update and define the connectedWallets list.
    // It uses walletConnector to get the current wallet address and network
    // The address and network in the runtime are updated by the event listeners
    // The event listeners are updating `connectedWallets` list.
    // It should be called on the first render and when connectedWalletsInfo updates
    const updateConnectedWalletsList = useCallback(() => __awaiter(void 0, void 0, void 0, function* () {
        const walletConnectors = walletConnectorOptions.map((wallet) => wallet.walletConnector);
        if (!walletConnectors.length) {
            return;
        }
        const updatedConnectedWallets = (yield Promise.all(connectedWalletsInfo.map((storedConnectedWalletInfo) => __awaiter(void 0, void 0, void 0, function* () {
            const walletConnector = getWalletConnectorByKey(walletConnectors, storedConnectedWalletInfo.walletConnectorKey);
            if (!walletConnector) {
                logger.error('Could not find walletConnector: ' +
                    storedConnectedWalletInfo.walletConnectorKey);
                yield disconnectWallet(storedConnectedWalletInfo.id);
                return null;
            }
            const [walletAddress] = yield walletConnector.getConnectedAccounts();
            if (!walletAddress) {
                yield disconnectWallet(storedConnectedWalletInfo.id);
                return null;
            }
            const [walletNetwork, additionalAddresses] = yield Promise.all([
                walletConnector.getNetwork(),
                walletConnector.getAdditionalAddresses(walletAddress),
            ]);
            const walletChain = walletConnector.connectedChain;
            const walletObject = {
                additionalAddresses,
                address: walletAddress || '',
                authenticated: false,
                chain: walletChain || '',
                connected: true,
                connector: walletConnector,
                id: storedConnectedWalletInfo.id,
                network: walletNetwork,
            };
            return walletObject;
        })))).filter((wallet) => Boolean(wallet));
        setConnectedWallets(updatedConnectedWallets);
        // Raise any onConnect callbacks in wait list
        for (const walletId of [...onConnectWaitList.current]) {
            const wallet = updatedConnectedWallets.find((wallet) => getWalletIdentifier(wallet) === walletId);
            if (!wallet)
                continue;
            onConnectWaitList.current.delete(walletId);
            dynamicEvents.emit('walletConnected', wallet);
            setIsVerificationInProgress(false);
        }
    }), [
        connectedWalletsInfo,
        setIsVerificationInProgress,
        disconnectWallet,
        walletConnectorOptions,
    ]);
    // Generate the connectedWallets list when connectedWalletsInfo (localStorage) or memoized wallet connector updates
    useEffect(() => {
        updateConnectedWalletsList();
        // eslint-disable-next-line react-hooks/exhaustive-deps
    }, [connectedWalletsInfo, walletConnectorOptions]);
    const getNextIdAvailable = useCallback((currentWallets) => {
        const existingIds = currentWallets.map(({ id }) => id);
        const sortedIds = existingIds
            .map((id) => Number(id.replace('connect-wallet-', '')))
            .sort((a, b) => a - b);
        for (let i = 0; i < sortedIds.length; i++) {
            if (sortedIds[i] !== i) {
                return i;
            }
        }
        return sortedIds.length;
    }, []);
    const updateConnectedWalletById = useCallback((walletId, walletObject) => {
        const clonedConnectedWalletsList = [...connectedWallets];
        const connectedWalletIndex = clonedConnectedWalletsList.findIndex((connectedWallet) => connectedWallet.id === walletId);
        if (connectedWalletIndex > -1) {
            if (walletObject.address && walletObject.network) {
                clonedConnectedWalletsList[connectedWalletIndex] = Object.assign(Object.assign({}, clonedConnectedWalletsList[connectedWalletIndex]), { address: walletObject.address, network: walletObject.network });
            }
            else if (walletObject.address) {
                clonedConnectedWalletsList[connectedWalletIndex] = Object.assign(Object.assign({}, clonedConnectedWalletsList[connectedWalletIndex]), { address: walletObject.address });
            }
            else {
                clonedConnectedWalletsList[connectedWalletIndex] = Object.assign(Object.assign({}, clonedConnectedWalletsList[connectedWalletIndex]), { network: walletObject.network });
            }
            setConnectedWallets(clonedConnectedWalletsList);
        }
    }, [connectedWallets]);
    // Keeps connected wallet data inside localStorage
    const applyConnectedWalletToStore = ({ walletId, walletConnectorKey, walletChain, provider, }) => {
        const updatedConnectedWalletsInfo = [...connectedWalletsInfo];
        updatedConnectedWalletsInfo.push({
            id: walletId,
            provider,
            walletChain,
            walletConnectorKey,
        });
        setConnectedWalletsInfo(updatedConnectedWalletsInfo);
    };
    // Updates connected wallet (network and address) and verifies connected wallet connection
    const refreshConnectedWallet = (walletId, walletConnector) => __awaiter(void 0, void 0, void 0, function* () {
        if (shouldManuallyReconnectOnRefresh(walletConnector)) {
            yield walletConnector.connect();
        }
        const currentWalletConnectorAddress = yield walletConnector.fetchPublicAddress();
        if (currentWalletConnectorAddress && onBeforeConnectSuccessConfirmation) {
            const shouldProceedWithConnection = yield onBeforeConnectSuccessConfirmation({
                address: currentWalletConnectorAddress,
                chain: walletConnector.connectedChain,
                connector: walletConnector,
            });
            if (!shouldProceedWithConnection) {
                logger.info('Connection was not established because onBeforeConnectSuccessConfirmation returned false');
                disconnectWallet(walletId);
                setShowAuthFlow(false, { emitCancelAuth: true });
                return;
            }
        }
        const currentWalletConnectorNetwork = yield walletConnector.getNetwork();
        updateConnectedWalletById(walletId, {
            address: currentWalletConnectorAddress,
            network: currentWalletConnectorNetwork,
        });
    });
    const shouldSkipCreatingVisit = authMode === 'connect-only' && !enableVisitTrackingOnConnectOnly;
    const connectWallet = (walletConnector, fetchPublicAddressOpts, options) => __awaiter(void 0, void 0, void 0, function* () {
        const { applyOnBeforeConnectSuccessConfirmation = true, raiseOnConnect = false, } = options !== null && options !== void 0 ? options : {};
        const walletAddress = yield walletConnector.fetchPublicAddress(fetchPublicAddressOpts);
        if (!walletAddress) {
            throw new MissingPublicAddressError();
        }
        if (onBeforeConnectSuccessConfirmation &&
            applyOnBeforeConnectSuccessConfirmation) {
            const shouldProceedWithConnection = yield onBeforeConnectSuccessConfirmation({
                address: walletAddress,
                chain: walletConnector.connectedChain,
                connector: walletConnector,
            });
            if (!shouldProceedWithConnection) {
                logger.info('Connection was not established because onBeforeConnectSuccessConfirmation returned false');
                setShowAuthFlow(false);
                return;
            }
        }
        const walletId = `connect-wallet-${getNextIdAvailable(connectedWalletsInfo)}`;
        // We are already using the new syntax for wallet ids for this.
        // Eventually the connect-wallet-X syntax will be dropped
        if (raiseOnConnect && isConnectOnly(user, authMode))
            onConnectWaitList.current.add(getWalletIdentifier({
                address: walletAddress,
                connector: walletConnector,
            }));
        if (!primaryWalletId) {
            setPrimaryWalletId(walletId);
        }
        const walletChain = walletConnector.connectedChain;
        const isWalletStored = connectedWalletsInfo.some(({ walletConnectorKey: storedWalletConnectorKey }) => storedWalletConnectorKey === walletConnector.key);
        if (isWalletStored) {
            yield updateConnectedWalletsList();
        }
        else {
            applyConnectedWalletToStore({
                provider: getWalletProvider(walletConnector),
                walletAddress: walletAddress,
                walletChain,
                walletConnectorKey: walletConnector.key,
                walletId,
            });
            if (!shouldSkipCreatingVisit) {
                // send information to backend to kick off background jobs
                // so verify/sign on the next step could go by more quickly
                // this is async work, but DO NOT AWAIT
                createVisit({
                    authMode,
                    chain: walletChain || '',
                    environmentId,
                    publicWalletAddress: walletAddress,
                    walletName: walletConnector.key,
                    walletProvider: getWalletProvider(walletConnector),
                });
            }
        }
        return { address: walletAddress, id: walletId };
    });
    const getConnectedWalletById = useCallback((walletId) => connectedWallets.find((wallet) => wallet.id === walletId), [connectedWallets]);
    return {
        connectWallet,
        connectedWallets,
        connectedWalletsInfo,
        disconnectWallet,
        getConnectedWalletById,
        refreshConnectedWallet,
        removeConnectedWalletsInfo,
    };
};

export { useConnectWallet };
