// @zerodev/sdk/_esm/actions/paymaster/sponsorUserOperation.js
import { deepHexlify } from "permissionless/utils";
import { KERNEL_ADDRESSES } from "../../accounts/index.js";
/**
 * Returns paymasterAndData & updated gas parameters required to sponsor a userOperation.
 */
export const sponsorUserOperation = async (client, args) => {
    const response = await client.request({
        method: "zd_sponsorUserOperation",
        params: [
            {
                chainId: client.chain?.id,
                userOp: deepHexlify(args.userOperation),
                entryPointAddress: args.entryPoint ?? KERNEL_ADDRESSES.ENTRYPOINT_V0_6,
                gasTokenData: args.gasToken && {
                    tokenAddress: args.gasToken
                },
                shouldOverrideFee: args.shouldOverrideFee ?? false,
                shouldConsume: args.shouldConsume ?? true
            }
        ]
    });
    let result = {
        ...args.userOperation,
        paymasterAndData: response.paymasterAndData,
        preVerificationGas: BigInt(response.preVerificationGas),
        verificationGasLimit: BigInt(response.verificationGasLimit),
        callGasLimit: BigInt(response.callGasLimit)
    };
    if (response.maxFeePerGas !== undefined &&
        response.maxPriorityFeePerGas !== undefined) {
        result = {
            ...result,
            maxFeePerGas: BigInt(response.maxFeePerGas),
            maxPriorityFeePerGas: BigInt(response.maxPriorityFeePerGas)
        };
    }
    return result;
};
