// @zerodev/sdk/dist/cjs/kernel-zerodev/owner/passkey/getAutocompletePasskeyOwner.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.getAutocompletePasskeyOwner = void 0;
const webauthn_1 = require("@turnkey/http/dist/webauthn");
const axios_1 = __importDefault(require("axios"));
const utils_js_1 = require("./utils.js");
const constants_js_1 = require("../../constants.js");
async function getAutocompletePasskeyOwner({ projectId, apiUrl = constants_js_1.API_URL, }) {
    if (!PublicKeyCredential?.isConditionalMediationAvailable ||
        !PublicKeyCredential?.isConditionalMediationAvailable()) {
        return;
    }
    if (typeof window !== "undefined") {
        const challenge = (0, utils_js_1.generateRandomBuffer)();
        try {
            (0, utils_js_1.abortWebauthn)();
            const assertion = JSON.parse(await (0, webauthn_1.getWebAuthnAssertion)((0, utils_js_1.base64UrlEncode)(challenge), {
                mediation: "conditional",
                publicKey: {
                    rpId: window.location.hostname,
                    userVerification: "required",
                },
                signal: utils_js_1.abortController.controller.signal,
            }));
            const response = await axios_1.default.post(`${apiUrl}/projects/${projectId}/wallets/${assertion.credentialId}`, {
                challenge: (0, utils_js_1.base64UrlEncode)(challenge),
                assertion,
            });
            const { id, walletId, address } = await response.data;
            console.log("owner", id, walletId, address);
            const owner = {
                getAddress: async () => address,
                signMessage: (msg) => (0, utils_js_1.signMessage)(msg, id, walletId, projectId, assertion.credentialId, apiUrl),
                signTypedData: (params) => (0, utils_js_1.signTypedData)(params, id, walletId, projectId, assertion.credentialId, apiUrl),
            };
            return owner;
        }
        catch (e) {
            console.log(e);
        }
    }
    return;
}
exports.getAutocompletePasskeyOwner = getAutocompletePasskeyOwner;
