// @zerodev/sdk/_cjs/actions/paymaster/sponsorUserOperation.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.sponsorUserOperation = void 0;
const utils_1 = require("permissionless/utils");
const index_js_1 = require("../../accounts/index.js");
const sponsorUserOperation = async (client, args) => {
    const response = await client.request({
        method: "zd_sponsorUserOperation",
        params: [
            {
                chainId: client.chain?.id,
                userOp: (0, utils_1.deepHexlify)(args.userOperation),
                entryPointAddress: args.entryPoint ?? index_js_1.KERNEL_ADDRESSES.ENTRYPOINT_V0_6,
                gasTokenData: args.gasToken && {
                    tokenAddress: args.gasToken
                },
                shouldOverrideFee: args.shouldOverrideFee ?? false,
                shouldConsume: args.shouldConsume ?? true
            }
        ]
    });
    let result = {
        ...args.userOperation,
        paymasterAndData: response.paymasterAndData,
        preVerificationGas: BigInt(response.preVerificationGas),
        verificationGasLimit: BigInt(response.verificationGasLimit),
        callGasLimit: BigInt(response.callGasLimit)
    };
    if (response.maxFeePerGas !== undefined &&
        response.maxPriorityFeePerGas !== undefined) {
        result = {
            ...result,
            maxFeePerGas: BigInt(response.maxFeePerGas),
            maxPriorityFeePerGas: BigInt(response.maxPriorityFeePerGas)
        };
    }
    return result;
};
exports.sponsorUserOperation = sponsorUserOperation;
