// @dynamic-labs/viem-utils/src/lib/confirmationTransport/confirmationTransport.js
import { __awaiter } from '../../../_virtual/_tslib.js';
import { fromHex } from 'viem';
import { interceptTransport } from '../interceptTransport/interceptTransport.js';

const confirmationTransport = ({ transport, walletUiUtils, getAccounts, onPersonalSign, onSendTransaction, onSignTypedData, provider, }) => interceptTransport({
    getAccounts,
    onPersonalSign: onPersonalSign
        ? (args) => walletUiUtils.signMessage({
            handler: () => onPersonalSign(args),
            message: fromHex(args.message, 'string'),
        })
        : undefined,
    onSendTransaction: onSendTransaction
        ? (args) => walletUiUtils.sendTransaction({
            handler: (transaction) => __awaiter(void 0, void 0, void 0, function* () {
                return onSendTransaction(Object.assign(Object.assign({}, args), { transaction: transaction }));
            }),
            provider,
            transaction: args.transaction,
        })
        : undefined,
    onSignTypedData: onSignTypedData
        ? (args) => walletUiUtils.signMessage({
            handler: () => onSignTypedData(args),
            message: args.message,
        })
        : undefined,
    transport,
});

export { confirmationTransport };
