// @alchemy/aa-core/dist/esm/signer/local-account.js
import { isHex } from "viem";
import { mnemonicToAccount, privateKeyToAccount } from "viem/accounts";
export class LocalAccountSigner {
    constructor(owner) {
        Object.defineProperty(this, "owner", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: void 0
        });
        Object.defineProperty(this, "signMessage", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: (msg) => {
                if (typeof msg === "string" && !isHex(msg)) {
                    return this.owner.signMessage({
                        message: msg,
                    });
                }
                else {
                    return this.owner.signMessage({
                        message: {
                            raw: msg,
                        },
                    });
                }
            }
        });
        Object.defineProperty(this, "signTypedData", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: (params) => {
                return this.owner.signTypedData(params);
            }
        });
        Object.defineProperty(this, "getAddress", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: async () => {
                return this.owner.address;
            }
        });
        this.owner = owner;
    }
    static mnemonicToAccountSigner(key) {
        const owner = mnemonicToAccount(key);
        return new LocalAccountSigner(owner);
    }
    static privateKeyToAccountSigner(key) {
        const owner = privateKeyToAccount(key);
        return new LocalAccountSigner(owner);
    }
}
