// @zerodev/sdk/dist/cjs/kernel-zerodev/validator-provider/base.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ValidatorProvider = void 0;
const provider_js_1 = require("../provider.js");
const account_js_1 = require("../account.js");
const chains_1 = require("viem/chains");
const paymaster_js_1 = require("../middleware/paymaster.js");
class ValidatorProvider extends provider_js_1.ZeroDevProvider {
    constructor(params, validator) {
        let bundlerProvider = params.bundlerProvider;
        const shouldUsePaymaster = (params.usePaymaster === undefined || params.usePaymaster) &&
            bundlerProvider !== "GELATO";
        if (params.opts?.paymasterConfig &&
            params.opts?.paymasterConfig.policy === "TOKEN_PAYMASTER" &&
            shouldUsePaymaster) {
            bundlerProvider = "STACKUP";
        }
        super({
            ...params.opts?.providerConfig,
            chain: params.opts?.providerConfig?.chain ?? chains_1.polygonMumbai,
            rpcUrl: params.opts?.providerConfig?.rpcUrl,
            projectId: params.projectId,
            bundlerProvider,
        });
        Object.defineProperty(this, "defaultProvider", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: void 0
        });
        Object.defineProperty(this, "getValidator", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: () => {
                if (!(0, account_js_1.isKernelAccount)(this.account) || !this.account.validator) {
                    throw new Error("ValidatorProvider: account with validator is not set, did you call all connects first?");
                }
                return this.account.getValidator();
            }
        });
        Object.defineProperty(this, "getEncodedEnableData", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: async (enableData) => {
                if (!(0, account_js_1.isKernelAccount)(this.account) || !this.account.validator) {
                    throw new Error("ValidatorProvider: account with validator is not set, did you call all connects first?");
                }
                return await this.account.validator.encodeEnable(enableData);
            }
        });
        Object.defineProperty(this, "getEncodedDisableData", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: async (disableData = "0x") => {
                if (!(0, account_js_1.isKernelAccount)(this.account) || !this.account.validator) {
                    throw new Error("ValidatorProvider: account with validator is not set, did you call all connects first?");
                }
                return await this.account.validator.encodeDisable(disableData);
            }
        });
        Object.defineProperty(this, "sendEnableUserOperation", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: async (enableData) => {
                const encodedEnableData = await this.getEncodedEnableData(enableData);
                if (!(0, account_js_1.isKernelAccount)(this.account) || !this.account.validator) {
                    throw new Error("ValidatorProvider: account with validator is not set, did you call all connects first?");
                }
                return await this.sendUserOperation({
                    target: this.account.validator.validatorAddress,
                    data: encodedEnableData,
                });
            }
        });
        Object.defineProperty(this, "sendDisableUserOperation", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: async (disableData = "0x") => {
                const encodedDisableData = await this.getEncodedDisableData(disableData);
                if (!(0, account_js_1.isKernelAccount)(this.account) || !this.account.validator) {
                    throw new Error("ValidatorProvider: account with validator is not set, did you call all connects first?");
                }
                if (!this.defaultProvider) {
                    throw Error("Default Validator provider unintialized");
                }
                return await this.defaultProvider.sendUserOperation({
                    target: this.account.validator.validatorAddress,
                    data: encodedDisableData,
                });
            }
        });
        this.defaultProvider = params.defaultProvider;
        if ((typeof params.opts?.providerConfig?.chain === "number" &&
            typeof params.defaultProvider?.chain === "number" &&
            params.opts?.providerConfig?.chain !== params.defaultProvider?.chain) ||
            (typeof params.opts?.providerConfig?.chain === "object" &&
                typeof params.defaultProvider?.chain === "object" &&
                params.opts?.providerConfig?.chain.id !==
                    params.defaultProvider?.chain.id)) {
            throw new Error("chainId mismatch between default provider and current provider - check projectIds");
        }
        this.connect(() => new account_js_1.KernelSmartContractAccount({
            projectId: params.projectId,
            validator,
            defaultValidator: params.defaultProvider?.getValidator(),
            rpcClient: this.rpcClient,
            bundlerProvider,
            index: params.defaultProvider?.getAccount().getIndex(),
            ...params.opts?.accountConfig,
        }));
        if (shouldUsePaymaster) {
            let paymasterConfig = params.opts?.paymasterConfig ?? {
                policy: "VERIFYING_PAYMASTER",
            };
            paymasterConfig = {
                ...paymasterConfig,
                paymasterProvider: params.opts?.paymasterConfig?.paymasterProvider ?? bundlerProvider,
            };
            (0, paymaster_js_1.withZeroDevPaymasterAndData)(this, paymasterConfig);
        }
    }
}
exports.ValidatorProvider = ValidatorProvider;
