// @zerodev/sdk/dist/cjs/kernel-zerodev/validator-provider/ecdsa-provider.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ECDSAProvider = void 0;
const base_js_1 = require("./base.js");
const ecdsa_validator_js_1 = require("../validator/ecdsa-validator.js");
const index_js_1 = require("../api/index.js");
const chains_1 = require("viem/chains");
const utils_js_1 = require("../utils.js");
class ECDSAProvider extends base_js_1.ValidatorProvider {
    constructor(params) {
        const chain = typeof params.opts?.providerConfig?.chain === "number"
            ? (0, utils_js_1.getChain)(params.opts.providerConfig.chain)
            : params.opts?.providerConfig?.chain ?? chains_1.polygonMumbai;
        const validator = new ecdsa_validator_js_1.ECDSAValidator({
            projectId: params.projectId,
            owner: params.owner,
            chain,
            rpcUrl: params.opts?.providerConfig?.rpcUrl,
            ...params.opts?.validatorConfig,
        });
        super({
            ...params,
            opts: {
                ...params.opts,
                providerConfig: { ...params.opts?.providerConfig, chain },
            },
        }, validator);
        Object.defineProperty(this, "changeOwner", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: this.sendEnableUserOperation
        });
        Object.defineProperty(this, "deleteOwner", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: this.sendDisableUserOperation
        });
    }
    static async init(params) {
        const chainId = await (0, index_js_1.getChainId)(params.projectId);
        if (!chainId) {
            throw new Error("ChainId not found");
        }
        const chain = (0, utils_js_1.getChain)(chainId);
        const instance = new ECDSAProvider({
            ...params,
            opts: {
                ...params.opts,
                providerConfig: {
                    chain,
                    ...params.opts?.providerConfig,
                },
            },
        });
        return instance;
    }
}
exports.ECDSAProvider = ECDSAProvider;
