// @zerodev/sdk/dist/cjs/kernel-zerodev/validator-provider/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.RecoveryProvider = exports.SessionKeyProvider = exports.ERC165SessionKeyProvider = exports.KillSwitchProvider = exports.ECDSAProvider = exports.ValidatorProviders = void 0;
const ecdsa_provider_js_1 = require("./ecdsa-provider.js");
Object.defineProperty(exports, "ECDSAProvider", { enumerable: true, get: function () { return ecdsa_provider_js_1.ECDSAProvider; } });
const erc165_session_key_provider_js_1 = require("./erc165-session-key-provider.js");
Object.defineProperty(exports, "ERC165SessionKeyProvider", { enumerable: true, get: function () { return erc165_session_key_provider_js_1.ERC165SessionKeyProvider; } });
const kill_switch_provider_js_1 = require("./kill-switch-provider.js");
Object.defineProperty(exports, "KillSwitchProvider", { enumerable: true, get: function () { return kill_switch_provider_js_1.KillSwitchProvider; } });
const recovery_provider_js_1 = require("./recovery-provider.js");
Object.defineProperty(exports, "RecoveryProvider", { enumerable: true, get: function () { return recovery_provider_js_1.RecoveryProvider; } });
const session_key_provider_js_1 = require("./session-key-provider.js");
Object.defineProperty(exports, "SessionKeyProvider", { enumerable: true, get: function () { return session_key_provider_js_1.SessionKeyProvider; } });
exports.ValidatorProviders = {
    ECDSA: ecdsa_provider_js_1.ECDSAProvider,
    KILL_SWITCH: kill_switch_provider_js_1.KillSwitchProvider,
    ERC165_SESSION_KEY: erc165_session_key_provider_js_1.ERC165SessionKeyProvider,
    SESSION_KEY: session_key_provider_js_1.SessionKeyProvider,
    RECOVERY: recovery_provider_js_1.RecoveryProvider,
};
