// @zerodev/sdk/dist/cjs/kernel-zerodev/validator-provider/kill-switch-provider.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.KillSwitchProvider = void 0;
const base_js_1 = require("./base.js");
const kill_switch_validator_js_1 = require("../validator/kill-switch-validator.js");
const index_js_1 = require("../api/index.js");
const chains_1 = require("viem/chains");
const constants_js_1 = require("../constants.js");
const utils_js_1 = require("../utils.js");
class KillSwitchProvider extends base_js_1.ValidatorProvider {
    constructor(params) {
        const chain = typeof params.opts?.providerConfig?.chain === "number"
            ? (0, utils_js_1.getChain)(params.opts.providerConfig.chain)
            : params.opts?.providerConfig?.chain ?? chains_1.polygonMumbai;
        const validator = new kill_switch_validator_js_1.KillSwitchValidator({
            projectId: params.projectId,
            guardian: params.guardian,
            delaySeconds: params.delaySeconds,
            chain,
            rpcUrl: params.opts?.providerConfig?.rpcUrl,
            validatorAddress: params.opts?.validatorConfig?.validatorAddress ??
                constants_js_1.KILL_SWITCH_VALIDATOR_ADDRESS,
            ...params.opts?.validatorConfig,
        });
        super({
            ...params,
            opts: {
                ...params.opts,
                providerConfig: { ...params.opts?.providerConfig, chain },
            },
        }, validator);
        Object.defineProperty(this, "changeGuardian", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: this.sendEnableUserOperation
        });
        Object.defineProperty(this, "deleteKillSwitchData", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: this.sendDisableUserOperation
        });
    }
    static async init(params) {
        const chainId = await (0, index_js_1.getChainId)(params.projectId);
        if (!chainId) {
            throw new Error("ChainId not found");
        }
        const chain = (0, utils_js_1.getChain)(chainId);
        const instance = new KillSwitchProvider({
            ...params,
            opts: {
                ...params.opts,
                providerConfig: {
                    chain,
                    ...params.opts?.providerConfig,
                },
            },
        });
        return instance;
    }
}
exports.KillSwitchProvider = KillSwitchProvider;
