// @dynamic-labs/sdk-api/src/apis/CustomFieldsApi.cjs
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var _tslib = require('../../_virtual/_tslib.cjs');
var runtime = require('../runtime.cjs');
require('../models/AccessOutcomeEnum.cjs');
require('../models/AuthModeEnum.cjs');
require('../models/AuthenticatorTransportProtocol.cjs');
require('../models/JwtVerifiedCredentialFormatEnum.cjs');
require('../models/ProviderEnum.cjs');
require('../models/WalletAddressType.cjs');
require('../models/HardwareWalletEnum.cjs');
require('../models/PasswordSourceTypeEnum.cjs');
require('../models/WalletProviderEnum.cjs');
require('../models/BillingSubscriptionPlanTypeEnum.cjs');
require('../models/ChainEnum.cjs');
require('../models/ChainalysisCheckResultEnum.cjs');
require('../models/EmbeddedWalletProviderEnum.cjs');
require('../models/UserIdentifierTypeEnum.cjs');
require('../models/CurrencyType.cjs');
var CustomField = require('../models/CustomField.cjs');
var CustomFieldRequest = require('../models/CustomFieldRequest.cjs');
require('../models/CustomFieldType.cjs');
require('../models/TimeUnitEnum.cjs');
require('../models/EmbeddedWalletAuthType.cjs');
require('../models/EmbeddedWalletSecurityMethod.cjs');
require('../models/EnvironmentEnum.cjs');
require('../models/SignInProviderEnum.cjs');
require('../models/SocialSignInProviderEnum.cjs');
require('../models/SdkViewSectionAlignment.cjs');
require('../models/SdkViewSectionType.cjs');
require('../models/SdkViewType.cjs');
require('../models/UserFilterableFieldsEnum.cjs');
require('../models/ExportFormatEnum.cjs');
require('../models/ExportModelEnum.cjs');
require('../models/ExportStatusEnum.cjs');
require('../models/GateRuleType.cjs');
require('../models/HealthcheckStatus.cjs');
require('../models/InviteStatusEnum.cjs');
require('../models/NextViewEnum.cjs');
require('../models/OauthResultStatus.cjs');
require('../models/WhenToImplementEnum.cjs');
require('../models/PublishEventsEvents.cjs');
require('../models/RoleEnum.cjs');
require('../models/SubscriptionAdvancedScopeEnum.cjs');
require('../models/SubscriptionFreeScopeEnum.cjs');
require('../models/UnprocessableEntityErrorCode.cjs');
require('../models/VisitorFilterableFieldsEnum.cjs');

/* tslint:disable */
/**
 *
 */
class CustomFieldsApi extends runtime.BaseAPI {
    /**
     * Create a new custom field for an environment
     */
    createCustomFieldForEnvironmentRaw(requestParameters, initOverrides) {
        return _tslib.__awaiter(this, void 0, void 0, function* () {
            if (requestParameters.environmentId === null || requestParameters.environmentId === undefined) {
                throw new runtime.RequiredError('environmentId', 'Required parameter requestParameters.environmentId was null or undefined when calling createCustomFieldForEnvironment.');
            }
            if (requestParameters.customFieldRequest === null || requestParameters.customFieldRequest === undefined) {
                throw new runtime.RequiredError('customFieldRequest', 'Required parameter requestParameters.customFieldRequest was null or undefined when calling createCustomFieldForEnvironment.');
            }
            const queryParameters = {};
            const headerParameters = {};
            headerParameters['Content-Type'] = 'application/json';
            if (this.configuration && this.configuration.accessToken) {
                const token = this.configuration.accessToken;
                const tokenString = yield token("bearerAuth", []);
                if (tokenString) {
                    headerParameters["Authorization"] = `Bearer ${tokenString}`;
                }
            }
            const response = yield this.request({
                path: `/environments/{environmentId}/custom/fields`.replace(`{${"environmentId"}}`, encodeURIComponent(String(requestParameters.environmentId))),
                method: 'POST',
                headers: headerParameters,
                query: queryParameters,
                body: CustomFieldRequest.CustomFieldRequestToJSON(requestParameters.customFieldRequest),
            }, initOverrides);
            return new runtime.JSONApiResponse(response, (jsonValue) => CustomField.CustomFieldFromJSON(jsonValue));
        });
    }
    /**
     * Create a new custom field for an environment
     */
    createCustomFieldForEnvironment(requestParameters, initOverrides) {
        return _tslib.__awaiter(this, void 0, void 0, function* () {
            const response = yield this.createCustomFieldForEnvironmentRaw(requestParameters, initOverrides);
            return yield response.value();
        });
    }
    /**
     * Delete a specific custom field by its ID
     */
    deleteCustomFieldByIdRaw(requestParameters, initOverrides) {
        return _tslib.__awaiter(this, void 0, void 0, function* () {
            if (requestParameters.customFieldId === null || requestParameters.customFieldId === undefined) {
                throw new runtime.RequiredError('customFieldId', 'Required parameter requestParameters.customFieldId was null or undefined when calling deleteCustomFieldById.');
            }
            const queryParameters = {};
            const headerParameters = {};
            if (this.configuration && this.configuration.accessToken) {
                const token = this.configuration.accessToken;
                const tokenString = yield token("bearerAuth", []);
                if (tokenString) {
                    headerParameters["Authorization"] = `Bearer ${tokenString}`;
                }
            }
            const response = yield this.request({
                path: `/custom/fields/{customFieldId}`.replace(`{${"customFieldId"}}`, encodeURIComponent(String(requestParameters.customFieldId))),
                method: 'DELETE',
                headers: headerParameters,
                query: queryParameters,
            }, initOverrides);
            return new runtime.VoidApiResponse(response);
        });
    }
    /**
     * Delete a specific custom field by its ID
     */
    deleteCustomFieldById(requestParameters, initOverrides) {
        return _tslib.__awaiter(this, void 0, void 0, function* () {
            yield this.deleteCustomFieldByIdRaw(requestParameters, initOverrides);
        });
    }
    /**
     * Retrieve a specific custom field by its ID
     */
    getCustomFieldByIdRaw(requestParameters, initOverrides) {
        return _tslib.__awaiter(this, void 0, void 0, function* () {
            if (requestParameters.customFieldId === null || requestParameters.customFieldId === undefined) {
                throw new runtime.RequiredError('customFieldId', 'Required parameter requestParameters.customFieldId was null or undefined when calling getCustomFieldById.');
            }
            const queryParameters = {};
            const headerParameters = {};
            if (this.configuration && this.configuration.accessToken) {
                const token = this.configuration.accessToken;
                const tokenString = yield token("bearerAuth", []);
                if (tokenString) {
                    headerParameters["Authorization"] = `Bearer ${tokenString}`;
                }
            }
            const response = yield this.request({
                path: `/custom/fields/{customFieldId}`.replace(`{${"customFieldId"}}`, encodeURIComponent(String(requestParameters.customFieldId))),
                method: 'GET',
                headers: headerParameters,
                query: queryParameters,
            }, initOverrides);
            return new runtime.JSONApiResponse(response, (jsonValue) => CustomField.CustomFieldFromJSON(jsonValue));
        });
    }
    /**
     * Retrieve a specific custom field by its ID
     */
    getCustomFieldById(requestParameters, initOverrides) {
        return _tslib.__awaiter(this, void 0, void 0, function* () {
            const response = yield this.getCustomFieldByIdRaw(requestParameters, initOverrides);
            return yield response.value();
        });
    }
    /**
     * Get the custom fields for an environment
     */
    getCustomFieldsForEnvironmentRaw(requestParameters, initOverrides) {
        return _tslib.__awaiter(this, void 0, void 0, function* () {
            if (requestParameters.environmentId === null || requestParameters.environmentId === undefined) {
                throw new runtime.RequiredError('environmentId', 'Required parameter requestParameters.environmentId was null or undefined when calling getCustomFieldsForEnvironment.');
            }
            const queryParameters = {};
            const headerParameters = {};
            if (this.configuration && this.configuration.accessToken) {
                const token = this.configuration.accessToken;
                const tokenString = yield token("bearerAuth", []);
                if (tokenString) {
                    headerParameters["Authorization"] = `Bearer ${tokenString}`;
                }
            }
            const response = yield this.request({
                path: `/environments/{environmentId}/custom/fields`.replace(`{${"environmentId"}}`, encodeURIComponent(String(requestParameters.environmentId))),
                method: 'GET',
                headers: headerParameters,
                query: queryParameters,
            }, initOverrides);
            return new runtime.JSONApiResponse(response, (jsonValue) => jsonValue.map(CustomField.CustomFieldFromJSON));
        });
    }
    /**
     * Get the custom fields for an environment
     */
    getCustomFieldsForEnvironment(requestParameters, initOverrides) {
        return _tslib.__awaiter(this, void 0, void 0, function* () {
            const response = yield this.getCustomFieldsForEnvironmentRaw(requestParameters, initOverrides);
            return yield response.value();
        });
    }
    /**
     * Update an existing custom field by its ID
     */
    updateCustomFieldByIdRaw(requestParameters, initOverrides) {
        return _tslib.__awaiter(this, void 0, void 0, function* () {
            if (requestParameters.customFieldId === null || requestParameters.customFieldId === undefined) {
                throw new runtime.RequiredError('customFieldId', 'Required parameter requestParameters.customFieldId was null or undefined when calling updateCustomFieldById.');
            }
            if (requestParameters.customFieldRequest === null || requestParameters.customFieldRequest === undefined) {
                throw new runtime.RequiredError('customFieldRequest', 'Required parameter requestParameters.customFieldRequest was null or undefined when calling updateCustomFieldById.');
            }
            const queryParameters = {};
            const headerParameters = {};
            headerParameters['Content-Type'] = 'application/json';
            if (this.configuration && this.configuration.accessToken) {
                const token = this.configuration.accessToken;
                const tokenString = yield token("bearerAuth", []);
                if (tokenString) {
                    headerParameters["Authorization"] = `Bearer ${tokenString}`;
                }
            }
            const response = yield this.request({
                path: `/custom/fields/{customFieldId}`.replace(`{${"customFieldId"}}`, encodeURIComponent(String(requestParameters.customFieldId))),
                method: 'PUT',
                headers: headerParameters,
                query: queryParameters,
                body: CustomFieldRequest.CustomFieldRequestToJSON(requestParameters.customFieldRequest),
            }, initOverrides);
            return new runtime.JSONApiResponse(response, (jsonValue) => CustomField.CustomFieldFromJSON(jsonValue));
        });
    }
    /**
     * Update an existing custom field by its ID
     */
    updateCustomFieldById(requestParameters, initOverrides) {
        return _tslib.__awaiter(this, void 0, void 0, function* () {
            const response = yield this.updateCustomFieldByIdRaw(requestParameters, initOverrides);
            return yield response.value();
        });
    }
}

exports.CustomFieldsApi = CustomFieldsApi;
