// @dynamic-labs/sdk-react-core/src/lib/context/DynamicContext/DynamicContext.cjs
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var _tslib = require('../../../../_virtual/_tslib.cjs');
var jsxRuntime = require('react/jsx-runtime');
var React = require('react');
var reactI18next = require('react-i18next');
var iconic = require('@dynamic-labs/iconic');
var utils = require('@dynamic-labs/utils');
var walletBook = require('@dynamic-labs/wallet-book');
var walletConnectorCore = require('@dynamic-labs/wallet-connector-core');
var app = require('../../app.cjs');
require('../../events/dynamicEvents.cjs');
var decodeJwt = require('../../shared/utils/functions/decodeJwt/decodeJwt.cjs');
var ViewContext = require('../ViewContext/ViewContext.cjs');
var logger = require('../../shared/logger.cjs');
var localStorage$1 = require('../../utils/constants/localStorage.cjs');
require('../../utils/constants/colors.cjs');
var values = require('../../utils/constants/values.cjs');
var localStorage = require('../../shared/utils/classes/storage/localStorage.cjs');
require('viem');
require('@dynamic-labs/sdk-api');
var useGlobalLoading = require('../../shared/utils/hooks/useGlobalLoading/useGlobalLoading.cjs');
var useLocalStorage = require('../../shared/utils/hooks/useLocalStorage/useLocalStorage.cjs');
require('../../shared/consts/index.cjs');
var CaptchaContext = require('../CaptchaContext/CaptchaContext.cjs');
var ErrorContext = require('../ErrorContext/ErrorContext.cjs');
require('@dynamic-labs/multi-wallet');
var getAuthToken = require('../../utils/functions/getAuthToken/getAuthToken.cjs');
require('viem/chains');
var AccessDeniedContext = require('../AccessDeniedContext/AccessDeniedContext.cjs');
var AccountExistsContext = require('../AccountExistsContext/AccountExistsContext.cjs');
var api = require('../../data/api.cjs');
var EmailVerificationContext = require('../EmailVerificationContext/EmailVerificationContext.cjs');
require('react-dom');
var useFetchNonce = require('../../utils/hooks/useFetchNonce/useFetchNonce.cjs');
var useFetchProjectSettings = require('../../utils/hooks/useFetchProjectSettings/useFetchProjectSettings.cjs');
var ThemeContext = require('../ThemeContext/ThemeContext.cjs');
var useSetWalletConnectorVerifiedCredentials = require('../../utils/hooks/useSetWalletConnectorVerifiedCredentials/useSetWalletConnectorVerifiedCredentials.cjs');
require('@dynamic-labs/types');
require('yup');
require('../MockContext/MockContext.cjs');
require('../../utils/hooks/useUserUpdateRequest/useUpdateUser/useUpdateUser.cjs');
var UserFieldEditorContext = require('../UserFieldEditorContext/UserFieldEditorContext.cjs');
var useValidateSession = require('../../utils/hooks/useValidateSession/useValidateSession.cjs');
var useWalletConnectorOptions = require('../../utils/hooks/useWalletConnectorOptions/useWalletConnectorOptions.cjs');
var useWalletConnectors = require('../../utils/hooks/useWalletConnectors/useWalletConnectors.cjs');
var useWalletUiUtils = require('../../utils/hooks/useWalletUiUtils/useWalletUiUtils.cjs');
require('../../components/Alert/Alert.cjs');
require('../../components/ShadowDOM/ShadowDOM.cjs');
require('../../components/IconButton/IconButton.cjs');
require('../../components/InlineWidget/InlineWidget.cjs');
require('../../components/IsBrowser/IsBrowser.cjs');
require('../../components/MenuList/Dropdown/Dropdown.cjs');
require('../../components/Transition/ZoomTransition/ZoomTransition.cjs');
require('../../components/Transition/SlideInUpTransition/SlideInUpTransition.cjs');
require('../../components/Transition/OpacityTransition/OpacityTransition.cjs');
require('../../components/Popper/Popper/Popper.cjs');
require('../../components/Popper/PopperContext/PopperContext.cjs');
require('react-focus-lock');
require('qrcode');
require('formik');
var locale = require('../../locale/locale.cjs');
require('../../components/OverlayCard/OverlayCard.context.cjs');
var SyncPasskeyFlow = require('../../components/SyncPasskeyFlow/SyncPasskeyFlow.cjs');
var Toolkit = require('../../components/Toolkit/Toolkit.cjs');
require('../../components/PasskeyCreatedSuccessBanner/PasskeyCreatedSuccessBanner.cjs');
var PasskeyContext = require('../PasskeyContext/PasskeyContext.cjs');
require('../../views/WalletList/WalletList.cjs');
require('@hcaptcha/react-hcaptcha');
var LoadingContext = require('../LoadingContext/LoadingContext.cjs');
var SocialRedirectContext = require('../SocialRedirectContext/SocialRedirectContext.cjs');
var WalletGroupContext = require('../WalletGroupContext/WalletGroupContext.cjs');
require('../../widgets/DynamicBridgeWidget/views/WalletsView/components/SecondaryWallets/SecondaryWallets.cjs');
require('../FooterAnimationContext/index.cjs');
var ApiEndpoint = require('../../config/ApiEndpoint.cjs');
var getMissingChains = require('../../utils/functions/getMissingChains/getMissingChains.cjs');
var parseTheme = require('../../utils/functions/parseTheme/parseTheme.cjs');
var useMultiWalletWidgetState = require('../../utils/hooks/multiWallet/useMultiWalletWidgetState/useMultiWalletWidgetState.cjs');
var useClearWalletConnectSessions = require('../../utils/hooks/useClearWalletConnectSessions/useClearWalletConnectSessions.cjs');
var useConnectWallet = require('../../utils/hooks/useConnectWallet/useConnectWallet.cjs');
var useFetchWalletsForChainsMap = require('../../utils/hooks/useFetchWalletsForChainsMap/useFetchWalletsForChainsMap.cjs');
var useHandleUnlinkWallet = require('../../utils/hooks/useHandleUnlinkWallet/useHandleUnlinkWallet.cjs');
var usePrimaryWalletId = require('../../utils/hooks/usePrimaryWalletId/usePrimaryWalletId.cjs');
var useProcessNetworkConfigurations = require('../../utils/hooks/useProcessNetworkConfigurations/useProcessNetworkConfigurations.cjs');
var useRpcProviders = require('../../utils/hooks/useRpcProviders/useRpcProviders.cjs');
var useWalletEventListeners = require('../../utils/hooks/useWalletEventListeners/useWalletEventListeners.cjs');
var DynamicBridgeWidgetContext = require('../../widgets/DynamicBridgeWidget/context/DynamicBridgeWidgetContext/DynamicBridgeWidgetContext.cjs');
var DynamicWidgetContext = require('../../widgets/DynamicWidget/context/DynamicWidgetContext.cjs');
require('../../widgets/DynamicWidget/components/DynamicWidgetCard/DynamicWidgetCard.cjs');
require('../../components/UserProfile/parts/UserProfileField/components/VerifiedEmailIcon/VerifiedEmailIcon.cjs');
require('../../widgets/DynamicWidget/components/DynamicWidgetHeader/DynamicWidgetHeader.cjs');
require('@dynamic-labs/viem-utils');
require('../../views/TransactionConfirmationView/helpers/transactionErrorMessage.cjs');
require('../../utils/hooks/useTransactionWithGasPrice/useTransactionWithGasPrice.cjs');
require('../../widgets/DynamicWidget/views/ManagePasskeysWidgetView/PasskeyCard/PasskeyCard.cjs');
var FundingWidget = require('../../widgets/FundingWidget/FundingWidget.cjs');
var ConnectWithEmailOtpProvider = require('../ConnectWithEmailOtpContext/ConnectWithEmailOtpProvider.cjs');
require('../ConnectWithEmailOtpContext/connectWithEmailOtpContext.cjs');
var FundingContext = require('../FundingContext/FundingContext.cjs');
var SendBalanceContext = require('../SendBalanceContext/SendBalanceContext.cjs');
var UseNetworkValidation = require('../UseNetworkValidation/UseNetworkValidation.cjs');
var UserWalletsContext = require('../UserWalletsContext/UserWalletsContext.cjs');
var WalletContext = require('../WalletContext/WalletContext.cjs');
var PhantomRedirectContext = require('../PhantomRedirectContext/PhantomRedirectContext.cjs');
var handleDisconnectWallet = require('./handleDisconnectWallet/handleDisconnectWallet.cjs');
var helpers = require('./helpers/helpers.cjs');
var useDisplayOrderState = require('./hooks/useDisplayOrderState/useDisplayOrderState.cjs');
var useEmailLoginState = require('./hooks/useEmailLoginState/useEmailLoginState.cjs');
var useHandleLogout = require('./hooks/useHandleLogout/useHandleLogout.cjs');
var useNameService = require('./hooks/useNameService/useNameService.cjs');
var useSelectedWalletConnector = require('./hooks/useSelectedWalletConnector/useSelectedWalletConnector.cjs');
var useShowAuthFlow = require('./hooks/useShowAuthFlow/useShowAuthFlow.cjs');
var usePasskeySuccessPopup = require('./hooks/usePasskeySuccessPopup/usePasskeySuccessPopup.cjs');
var useTieCallbacksToEvents = require('./hooks/useTieCallbacksToEvents/useTieCallbacksToEvents.cjs');
var useVerificationInProgress = require('./hooks/useVerificationInProgress/useVerificationInProgress.cjs');
var useMultiWallet = require('./hooks/useMultiWallet/useMultiWallet.cjs');
var useSetShowLinkNewWalletModal = require('./hooks/useSetShowLinkNewWalletModal/useSetShowLinkNewWalletModal.cjs');
var useCustomerCallbacks = require('./useCustomerCallbacks/useCustomerCallbacks.cjs');
var validators = require('./validators/validators.cjs');
var useParentDynamicContextGuard = require('./hooks/useParentDynamicContextGuard/useParentDynamicContextGuard.cjs');

const PUBLIC_PROJECT_LIVE_ENVIRONMENT_ID = '2762a57b-faa4-41ce-9f16-abff9300e2c9';
const EMPTY_INITIAL_USER = undefined;
const DynamicContext = React.createContext(undefined);
/** The context provider itself we only use internally */
const InnerDynamicContextProvider = ({ children, theme, settings, locale: locale$1, }) => {
    var _a, _b, _c, _d, _e, _f, _g;
    const { accessDeniedMessagePrimary, accessDeniedMessageSecondary, accessDeniedButton, apiBaseUrl, enableConnectOnlyFallback = false, deepLinkPreference: deepLinkPreferenceProp = 'native', bridgeChains, cssOverrides, defaultNumberOfWalletsToShow = values.DEFAULT_NUMBER_OF_WALLETS_TO_SHOW, flowNetwork, initialAuthenticationMode = 'connect-and-sign', debugError = false, displaySiweStatement = true, newToWeb3WalletChainMap, enableVisitTrackingOnConnectOnly = true, environmentId, localStorageSuffix, eventsCallbacks, evmNetworks, walletsFilter, logLevel = 'WARN', mobileExperience = 'in-app-browser', notInTheListImageUrl, onboardingImageUrl, policiesConsentInnerComponent, customPrivacyPolicy, privacyPolicyUrl, signWithEmailWalletName, socialMediaLinkText, socialMediaIconUrl, socialMediaUrl, customTermsOfServices, termsOfServiceUrl, toolkitEnabled, siweStatement, shadowDOMEnabled = true, walletConnectors: walletConnectorsProp, socialProvidersFilter, showLockedWalletView = false, walletConnectPreferredChains, walletConnectorExtensions, hideEmbeddedWalletTransactionUIs, } = settings;
    const networkValidationMode = helpers.resolveNetworkValidationMode({
        bridgeChains: settings.bridgeChains,
        enableForcedNetworkValidation: settings.enableForcedNetworkValidation,
        networkValidationMode: settings.networkValidationMode,
    });
    let { appLogoUrl = 'https://demo.dynamic.xyz/assets/dynamic-logo.svg', appName = 'Dynamic Example', } = settings;
    const { walletConnectV1Bridge } = settings;
    if (!environmentId) {
        throw new utils.MissingEnvironmentIdError();
    }
    logger.logger.setLogLevel(logLevel);
    walletConnectorCore.logger.setLogLevel(logLevel);
    if (environmentId === PUBLIC_PROJECT_LIVE_ENVIRONMENT_ID) {
        logger.logger.warn(`WARNING: DYNAMIC is using a test environment ID ${environmentId}. Please sign up on https://app.dynamic.xyz/ to get your production environment ID.`);
    }
    ApiEndpoint.setBaseUrl((_a = process.env.DYNAMIC_API_BASE_URL) !== null && _a !== void 0 ? _a : apiBaseUrl);
    localStorage.LocalStorage.setSuffix(localStorageSuffix);
    const i18nSDKInstance = locale.Locale.setup(locale$1);
    const isAuthenticated = Boolean(getAuthToken.getAuthToken());
    const isBridgeFlow = Boolean(bridgeChains);
    const deepLinkPreference = helpers.getDeepLinkPreference(deepLinkPreferenceProp, isBridgeFlow);
    const [onboardingOnlyJwt, setOnboardingOnlyJwt] = React.useState(undefined);
    // state to track the full connectivity status of the individual's wallets.
    // Set to true once the predetermined number of wallets is connected.
    const [bridgeOnboardingCompleted, setBridgeOnboardingCompleted, removeBridgeOnboardingCompleted,] = useLocalStorage.useLocalStorage(localStorage$1.BRIDGE_ONBOARDING_COMPLETED, false);
    const [isFullyConnected, setIsFullyConnected] = React.useState(false);
    // Allows clients to control widget flow without need to click on AccountControl
    const [showDynamicUserProfile, setShowDynamicUserProfile] = React.useState(false);
    // Alows clients to control the Bridget Widget flow without need to click on Bridge Widget
    const [showBridgeWidget, setShowBridgeWidget] = React.useState(false);
    const [isSingleWalletAccount, setIsSingleWalletAccount] = React.useState(false);
    const [user, setUser, removeUser] = useLocalStorage.useLocalStorage(localStorage$1.AUTH_USER, EMPTY_INITIAL_USER, validators.validateAuthUser, [localStorage$1.AUTH_TOKEN]);
    const { consumeNonce } = useFetchNonce.useFetchNonce(environmentId);
    const [qrcodeUri, setQrcodeUri] = React.useState('');
    const [desktopUri, setDesktopUri] = React.useState('');
    const [multiWalletWidgetState, setMultiWalletWidgetState, { awaiting_account_switch: [accountSwitchState], awaiting_signature: [awaitingSignatureState], },] = useMultiWalletWidgetState.useMultiWalletWidgetState();
    const { projectSettings, isLoading: isProjectSettingsLoading, removeLsSettings, } = useFetchProjectSettings.useFetchProjectSettings({
        authToken: getAuthToken.getAuthToken(),
        dynamicContextProps: { settings },
        environmentId,
    });
    const multiWallet = useMultiWallet.useMultiWallet({
        isBridgeFlow,
        multiWalletOverride: (_b = settings.overrides) === null || _b === void 0 ? void 0 : _b.multiWallet,
        multiWalletSettings: (_c = projectSettings === null || projectSettings === void 0 ? void 0 : projectSettings.sdk) === null || _c === void 0 ? void 0 : _c.multiWallet,
    });
    // ensures that when user comes back to site and primary wallet is out of sync, we don't show widget popup
    const [showWidgetStatePopup, setShowWidgetStatePopup] = React.useState(!multiWallet);
    // Used inside useVerifyOnAwaitingSignature. Is legacy and should be removed asap
    const [legacyIsVerifying, setLegacyIsVerifying] = React.useState(false);
    // Set to true when a customer starts any auth verification process and set to false when it ends
    const [isVerificationInProgress, setIsVerificationInProgress] = useVerificationInProgress.useVerificationInProgress();
    const walletBook$1 = walletBook.useWalletBookCdn();
    const { configurations: serverNetworkConfigurations, handleRemoveLsNetworks, } = useProcessNetworkConfigurations.useProcessNetworkConfigurations({
        authToken: getAuthToken.getAuthToken(),
        environmentId,
        evmNetworks,
        projectSettings,
    });
    const { imageUserInAccessList, imageUserNotInAccessList, displayName, appLogo, } = (projectSettings === null || projectSettings === void 0 ? void 0 : projectSettings.general) || {};
    appLogoUrl = appLogo || appLogoUrl;
    appName = displayName || appName;
    const [loginWithEmail, setLogInWithEmail, resetLoginWithEmail] = useEmailLoginState.useEmailLoginState((projectSettings === null || projectSettings === void 0 ? void 0 : projectSettings.providers) || [], user);
    const [authorizationViewDisplayOrder, setAuthorizationViewDisplayOrder] = useDisplayOrderState.useDisplayOrderState(projectSettings);
    const [confirmationModal, walletUiUtils] = useWalletUiUtils.useWalletUiUtils({
        appLogoUrl,
        appName,
        getAppOrigin: () => window.location.origin,
        hideEmbeddedWalletTransactionUIs,
    });
    const { walletConnectorOptions } = useWalletConnectorOptions.useWalletConnectorOptions({
        appLogoUrl,
        appName,
        deepLinkPreference,
        flowNetwork,
        mobileExperience,
        networkConfigurations: serverNetworkConfigurations,
        projectSettings,
        walletBook: walletBook$1,
        walletConnectPreferredChains,
        walletConnectV1Bridge,
        walletConnectorExtensions,
        walletConnectorsProp,
        walletUiUtils,
    });
    const [authMode, setAuthMode] = useLocalStorage.useLocalStorage(localStorage$1.AUTH_MODE, initialAuthenticationMode);
    const { clearPrimaryWalletId, primaryWalletId, setPrimaryWalletId } = usePrimaryWalletId.usePrimaryWalletId();
    const isUserAuthenticated = Boolean(authMode === 'connect-and-sign' ? user : isFullyConnected);
    const isRenderingEmbeddedAuthFlow = React.useRef(false);
    const [showAuthFlow, setShowAuthFlow] = useShowAuthFlow.useShowAuthFlow({
        isAuthenticated: isUserAuthenticated,
        isMultiWalletEnabled: multiWallet,
        isRenderingEmbeddedAuthFlow,
        walletConnectorOptions,
    });
    const { connectWallet, removeConnectedWalletsInfo, connectedWalletsInfo, getConnectedWalletById, connectedWallets, disconnectWallet, refreshConnectedWallet, } = useConnectWallet.useConnectWallet({
        authMode,
        clearPrimaryWalletId,
        enableVisitTrackingOnConnectOnly,
        environmentId,
        isBridgeFlow,
        onBeforeConnectSuccessConfirmation: eventsCallbacks === null || eventsCallbacks === void 0 ? void 0 : eventsCallbacks.onBeforeConnectSuccessConfirmation,
        onDisconnect: (wallet) => handleDisconnectWallet.handleDisconnectWallet({
            eventsCallbacks,
            wallet,
        }),
        primaryWalletId,
        setIsVerificationInProgress,
        setPrimaryWalletId,
        setShowAuthFlow,
        user,
        walletConnectorOptions,
    });
    const bridgeChainsToConnect = React.useMemo(() => getMissingChains.getMissingChains(bridgeChains, connectedWalletsInfo), [bridgeChains, connectedWalletsInfo]);
    const { selectedWalletConnector, setSelectedWalletConnectorKey } = useSelectedWalletConnector.useSelectedWalletConnector({ walletConnectorOptions });
    //selected wallet to perform an action (become primary, unlink, etc)
    const [selectedWalletWithAction, setSelectedWalletWithAction] = React.useState(null);
    // until user completes onboarding, jwt is only stored in onboardingOnlyJwt
    const authToken = (_d = getAuthToken.getAuthToken()) !== null && _d !== void 0 ? _d : onboardingOnlyJwt;
    useSetWalletConnectorVerifiedCredentials.useSetWalletConnectorVerifiedCredentials(authToken, walletConnectorOptions);
    const { didConnectedStateLoad, primaryWallet, showQrcodeModal, secondaryWallets, setPrimaryWallet, setShowQrcodeModal, wallets: linkedOrConnectedWallets, } = useWalletConnectors.useWalletConnectors({
        authMode,
        authToken: getAuthToken.getAuthToken() || onboardingOnlyJwt,
        canHaveMultipleWalletsConnected: multiWallet,
        connectedWallets,
        multiWalletWidgetState,
        onboardingOnlyJwt,
        primaryWalletId,
        setDesktopUri,
        setMultiWalletWidgetState,
        setPrimaryWalletId,
        setQrcodeUri,
        user,
        walletConnectorOptions,
    });
    React.useEffect(() => {
        setIsFullyConnected(!(bridgeChainsToConnect === null || bridgeChainsToConnect === void 0 ? void 0 : bridgeChainsToConnect.length) && Boolean(primaryWallet));
    }, [bridgeChainsToConnect, primaryWallet]);
    const walletConnectors = linkedOrConnectedWallets.map((wallet) => wallet.connector);
    const { sdkHasLoaded, setSdkHasLoaded } = useGlobalLoading.useGlobalLoading({
        authMode,
        connectedInfo: connectedWalletsInfo[0],
        connectedWallets,
        primaryWallet,
        projectSettings,
        user,
        walletBook: walletBook$1,
    });
    const setShowLinkNewWalletModal = useSetShowLinkNewWalletModal.useSetShowLinkNewWalletModal({
        isAuthenticated: isUserAuthenticated,
        isMultiWalletEnabled: multiWallet,
        setShowAuthFlow,
    });
    const { getNameService, removeConnectedNameService } = useNameService.useNameService({
        authToken,
        currentWallet: primaryWallet !== null && primaryWallet !== void 0 ? primaryWallet : connectedWallets[0],
    });
    const { clearAllWalletConnectSessions } = useClearWalletConnectSessions.useClearWalletConnectSessions({
        connectors: walletConnectors,
    });
    const handleLogOut = useHandleLogout.useHandleLogout({
        bridgeChains,
        clearAllWalletConnectSessions,
        clearPrimaryWalletId,
        connectedWallets,
        environmentId,
        eventsCallbacks,
        handleRemoveLsNetworks,
        initialAuthenticationMode,
        isVerificationInProgress,
        removeBridgeOnboardingCompleted,
        removeConnectedNameService,
        removeConnectedWalletsInfo,
        removeLsSettings,
        removeUser,
        resetLoginWithEmail,
        setAuthMode,
        setIsVerificationInProgress,
        setLegacyIsVerifying,
        setMultiWalletWidgetState,
        setOnboardingOnlyJwt,
        setSdkHasLoaded,
        setShowAuthFlow,
        setShowBridgeWidget,
        setShowDynamicUserProfile,
        user,
        walletConnectors,
    });
    let wcVersion;
    if (projectSettings) {
        wcVersion =
            ((_f = (_e = projectSettings === null || projectSettings === void 0 ? void 0 : projectSettings.sdk) === null || _e === void 0 ? void 0 : _e.walletConnect) === null || _f === void 0 ? void 0 : _f.v2Enabled) && !walletConnectV1Bridge
                ? 2
                : 1;
    }
    const { setHasLoggedOut } = useValidateSession.useValidateSession({
        authToken,
        handleLogOut,
        user,
        walletConnectV1Bridge,
        walletConnectors,
        wcVersion,
    });
    // Ties all client callbacks to dynamic's events, such that
    // when our events are triggered, so are the client callbacks
    useTieCallbacksToEvents.useTieCallbacksToEvents({
        clientCallbacks: eventsCallbacks,
        connectedWallets,
        setHasLoggedOut,
        user,
    });
    const handleUnlinkWallet = useHandleUnlinkWallet.useHandleUnlinkWallet({
        environmentId,
        eventsCallbacks,
        primaryWalletId,
        secondaryWallets,
        setUser,
        user,
        verifiedCredentials: (user === null || user === void 0 ? void 0 : user.verifiedCredentials) || [],
    });
    const { loadingNetwork, network, setNetwork } = useWalletEventListeners.useWalletEventListeners({
        authMode,
        disconnectWallet,
        handleLogOut,
        multiWallet,
        multiWalletWidgetState,
        primaryWallet,
        refreshConnectedWallet,
        secondaryWallets,
        selectedWalletConnector,
        selectedWalletWithAction,
        setMultiWalletWidgetState,
        setPrimaryWalletId,
        setSelectedWalletConnectorKey,
        setSelectedWalletWithAction,
        user,
    });
    const { walletsForChainsMap } = useFetchWalletsForChainsMap.useFetchWalletsForChainsMap(newToWeb3WalletChainMap);
    const rpcProviders = useRpcProviders.useRpcProviders({
        networkConfigurations: serverNetworkConfigurations,
    });
    React.useEffect(() => {
        helpers.initExpirationTime(handleLogOut);
    }, [handleLogOut]);
    const { setCallback } = useCustomerCallbacks.useCustomerCallbacks({
        callbacks: {
            onAuthSuccess: eventsCallbacks === null || eventsCallbacks === void 0 ? void 0 : eventsCallbacks.onAuthSuccess,
            onEmbeddedWalletCreated: eventsCallbacks === null || eventsCallbacks === void 0 ? void 0 : eventsCallbacks.onEmbeddedWalletCreated,
            onLinkSuccess: eventsCallbacks === null || eventsCallbacks === void 0 ? void 0 : eventsCallbacks.onLinkSuccess,
            onUserProfileUpdate: eventsCallbacks === null || eventsCallbacks === void 0 ? void 0 : eventsCallbacks.onUserProfileUpdate,
        },
        handleLogOut,
        isAuthenticated,
        primaryWallet,
        secondaryWallets,
        setHasLoggedOut,
        setIsVerificationInProgress,
        user,
    });
    const sendWagmiSettings = React.useCallback(() => (settings) => {
        const { dynamicWagmiSettings } = settings;
        if (dynamicWagmiSettings) {
            api.sendDynamicProps(environmentId, {
                dynamicWagmiSettings: { dynamicWagmiSettings },
            });
        }
    }, [environmentId]);
    const clearStatesOnBackClick = React.useCallback(() => _tslib.__awaiter(void 0, void 0, void 0, function* () {
        setDesktopUri('');
        // Clear verification flag but dont require it to be set
        if (isVerificationInProgress)
            setIsVerificationInProgress(false);
        setLegacyIsVerifying(false);
        if (!user) {
            clearAllWalletConnectSessions();
        }
        setMultiWalletWidgetState('idle');
        if (authMode !== 'connect-only') {
            removeConnectedWalletsInfo();
        }
        if (!localStorage.LocalStorage.getFromLS(localStorage$1.AUTH_USER)) {
            localStorage.LocalStorage.removeFromLS(localStorage$1.AUTH_TOKEN);
        }
        if (localStorage.LocalStorage.getFromLS(localStorage$1.WALLET_PICKER_SEARCH_KEY)) {
            localStorage.LocalStorage.removeFromLS(localStorage$1.WALLET_PICKER_SEARCH_KEY);
        }
        if (selectedWalletConnector &&
            (!primaryWallet ||
                primaryWallet.connector.key !== selectedWalletConnector.key)) {
            yield (selectedWalletConnector === null || selectedWalletConnector === void 0 ? void 0 : selectedWalletConnector.endSession());
        }
    }), [
        isVerificationInProgress,
        setIsVerificationInProgress,
        user,
        setMultiWalletWidgetState,
        authMode,
        selectedWalletConnector,
        primaryWallet,
        clearAllWalletConnectSessions,
        removeConnectedWalletsInfo,
    ]);
    const { passkeySuccessPopup, setPasskeySuccessPopup } = usePasskeySuccessPopup.usePasskeySuccessPopup();
    const value = React.useMemo(() => {
        var _a, _b;
        return ({
            accessDeniedButton,
            accessDeniedMessagePrimary,
            accessDeniedMessageSecondary,
            accountSwitchState,
            appLogoUrl,
            appName,
            authMode,
            authToken: getAuthToken.getAuthToken(),
            authorizationViewDisplayOrder,
            awaitingSignatureState,
            bridgeChains,
            bridgeChainsToConnect,
            bridgeOnboardingCompleted,
            clearStatesOnBackClick,
            connectWallet,
            connectedWallets,
            consumeNonce,
            cssOverrides,
            customPrivacyPolicy,
            customTermsOfServices,
            debugError,
            decodedAuthToken: decodeJwt.decodeJwt(getAuthToken.getAuthToken()),
            decodedOnboardingOnlyToken: decodeJwt.decodeJwt(onboardingOnlyJwt),
            defaultNumberOfWalletsToShow,
            desktopUri,
            disconnectWallet,
            displaySiweStatement,
            enableConnectOnlyFallback,
            environmentId,
            eventsCallbacks,
            getConnectedWalletById,
            getNameService,
            handleLogOut,
            handleUnlinkWallet: authMode === 'connect-only' ? disconnectWallet : handleUnlinkWallet,
            i18nSDKInstance,
            isAuthenticated,
            isBridgeFlow,
            isFullyConnected,
            isProjectSettingsLoading,
            isRenderingEmbeddedAuthFlow,
            isSingleWalletAccount,
            isVerificationInProgress,
            legacyIsVerifying,
            linkedWallets: authMode === 'connect-only' || !user ? [] : linkedOrConnectedWallets,
            loadingNetwork,
            locale: {
                changeLanguage: locale.Locale.getInstance().changeLanguage,
            },
            loginWithEmail,
            multiWallet,
            multiWalletWidgetState,
            network,
            networkConfigurations: serverNetworkConfigurations,
            networkValidationMode,
            newToWeb3WalletChainMap: walletsForChainsMap,
            notInTheListImageUrl: imageUserNotInAccessList || notInTheListImageUrl,
            onboardingImageUrl: imageUserInAccessList || onboardingImageUrl,
            onboardingOnlyJwt,
            overrides: settings.overrides,
            passkeySuccessPopup,
            policiesConsentInnerComponent,
            primaryWallet,
            primaryWalletId,
            privacyPolicyUrl,
            projectSettings,
            qrcodeUri,
            removeBridgeOnboardingCompleted,
            removeConnectedWalletsInfo,
            rpcProviders,
            sdkHasLoaded,
            secondaryWallets,
            selectedWalletConnector,
            selectedWalletWithAction,
            sendWagmiSettings,
            setAuthMode,
            setAuthorizationViewDisplayOrder,
            setBridgeOnboardingCompleted,
            setCallback,
            setDesktopUri,
            setIsSingleWalletAccount,
            setIsVerificationInProgress,
            setLegacyIsVerifying,
            setLogInWithEmail,
            setMultiWalletWidgetState,
            setNetwork,
            setOnboardingOnlyJwt,
            setPasskeySuccessPopup,
            setPrimaryWallet,
            setPrimaryWalletId,
            setQrcodeUri,
            setSdkHasLoaded,
            setSelectedWalletConnectorKey,
            setSelectedWalletWithAction,
            setShowAuthFlow,
            setShowBridgeWidget,
            setShowDynamicUserProfile,
            setShowLinkNewWalletModal,
            setShowQrcodeModal,
            setShowWidgetStatePopup,
            setUser,
            shadowDOMEnabled,
            showAuthFlow,
            showBridgeWidget,
            showDynamicUserProfile,
            showLockedWalletView,
            showQrcodeModal,
            showWidgetStatePopup,
            signWithEmailWalletName,
            siweStatement,
            socialMediaIconUrl: ((_a = projectSettings === null || projectSettings === void 0 ? void 0 : projectSettings.general) === null || _a === void 0 ? void 0 : _a.supportUrls) || socialMediaIconUrl,
            socialMediaLinkText,
            socialMediaUrl,
            socialProvidersFilter,
            termsOfServiceUrl,
            theme,
            toolkitEnabled,
            user,
            walletConnector: (_b = primaryWallet === null || primaryWallet === void 0 ? void 0 : primaryWallet.connector) !== null && _b !== void 0 ? _b : null,
            walletConnectorOptions,
            walletUiUtils,
            walletsFilter,
        });
    }, [
        accessDeniedButton,
        accessDeniedMessagePrimary,
        accessDeniedMessageSecondary,
        accountSwitchState,
        appLogoUrl,
        appName,
        authMode,
        authorizationViewDisplayOrder,
        awaitingSignatureState,
        bridgeChains,
        bridgeChainsToConnect,
        bridgeOnboardingCompleted,
        clearStatesOnBackClick,
        connectWallet,
        connectedWallets,
        consumeNonce,
        cssOverrides,
        customPrivacyPolicy,
        customTermsOfServices,
        debugError,
        onboardingOnlyJwt,
        defaultNumberOfWalletsToShow,
        desktopUri,
        disconnectWallet,
        displaySiweStatement,
        enableConnectOnlyFallback,
        environmentId,
        eventsCallbacks,
        getConnectedWalletById,
        getNameService,
        handleLogOut,
        handleUnlinkWallet,
        i18nSDKInstance,
        isAuthenticated,
        isBridgeFlow,
        isFullyConnected,
        isProjectSettingsLoading,
        isSingleWalletAccount,
        isVerificationInProgress,
        legacyIsVerifying,
        user,
        linkedOrConnectedWallets,
        loadingNetwork,
        loginWithEmail,
        multiWallet,
        multiWalletWidgetState,
        network,
        serverNetworkConfigurations,
        networkValidationMode,
        walletsForChainsMap,
        imageUserNotInAccessList,
        notInTheListImageUrl,
        imageUserInAccessList,
        onboardingImageUrl,
        settings.overrides,
        passkeySuccessPopup,
        policiesConsentInnerComponent,
        primaryWallet,
        primaryWalletId,
        privacyPolicyUrl,
        projectSettings,
        qrcodeUri,
        removeBridgeOnboardingCompleted,
        removeConnectedWalletsInfo,
        rpcProviders,
        sdkHasLoaded,
        secondaryWallets,
        selectedWalletConnector,
        selectedWalletWithAction,
        sendWagmiSettings,
        setAuthMode,
        setAuthorizationViewDisplayOrder,
        setBridgeOnboardingCompleted,
        setCallback,
        setIsVerificationInProgress,
        setLogInWithEmail,
        setMultiWalletWidgetState,
        setNetwork,
        setPrimaryWallet,
        setPrimaryWalletId,
        setSdkHasLoaded,
        setSelectedWalletConnectorKey,
        setShowAuthFlow,
        setShowLinkNewWalletModal,
        setPasskeySuccessPopup,
        setShowQrcodeModal,
        setUser,
        shadowDOMEnabled,
        showAuthFlow,
        showBridgeWidget,
        showDynamicUserProfile,
        showLockedWalletView,
        showQrcodeModal,
        showWidgetStatePopup,
        signWithEmailWalletName,
        siweStatement,
        socialMediaIconUrl,
        socialMediaLinkText,
        socialMediaUrl,
        socialProvidersFilter,
        termsOfServiceUrl,
        theme,
        toolkitEnabled,
        walletConnectorOptions,
        walletUiUtils,
        walletsFilter,
    ]);
    React.useEffect(() => {
        const addPrefetch = (attr) => {
            if (!document.head.querySelector(`#${attr.id}`)) {
                const element = document.createElement('link');
                element.setAttribute('id', attr.id);
                element.setAttribute('rel', 'prefetch');
                element.setAttribute('href', attr.href);
                element.setAttribute('as', attr.as);
                element.setAttribute('type', attr.type);
                document.head.insertAdjacentElement('beforeend', element);
            }
        };
        addPrefetch({
            as: 'image',
            href: iconic.getIconicSpriteUrl(),
            id: 'sprite',
            type: 'image/svg+xml',
        });
        addPrefetch({
            as: 'fetch',
            href: walletBook.getWalletBookCdnUrl(),
            id: 'wallet-book',
            type: 'application/json',
        });
    }, []);
    // DYN-1140 - Opens Widget when multiWalletWidgetState changes
    React.useEffect(() => {
        // If statement to secure SingleWallet widget
        if (!multiWallet && multiWalletWidgetState !== 'awaiting_account_switch') {
            return;
        }
        // We need to make sure that the user is logged in so that we do not try to display the widget before authorization
        if (user && multiWalletWidgetState !== 'idle' && !showDynamicUserProfile) {
            setShowDynamicUserProfile(true);
        }
    }, [multiWallet, multiWalletWidgetState, showDynamicUserProfile, user]);
    return (jsxRuntime.jsx(reactI18next.I18nextProvider, { i18n: i18nSDKInstance, children: jsxRuntime.jsx(DynamicContext.Provider, { value: value, children: jsxRuntime.jsx(walletBook.WalletBookContextProvider, { walletBook: walletBook$1, children: jsxRuntime.jsx(ThemeContext.ThemeContextProvider, { customerTheme: parseTheme.parseTheme(theme, ((_g = projectSettings === null || projectSettings === void 0 ? void 0 : projectSettings.design) === null || _g === void 0 ? void 0 : _g.modal) || undefined), designSettings: projectSettings === null || projectSettings === void 0 ? void 0 : projectSettings.design, children: jsxRuntime.jsx(LoadingContext.LoadingContextProvider, { children: jsxRuntime.jsxs(ViewContext.ViewContextProvider, { initialView: helpers.getInitialView({
                                bridgeOnboardingCompleted,
                                connectedWallets,
                                isAuthenticated: isAuthenticated || Boolean(primaryWallet),
                                isBridgeFlow,
                                isMultiWalletEnabled: multiWallet,
                            }), children: [networkValidationMode === 'always' && jsxRuntime.jsx(UseNetworkValidation.UseNetworkValidation, {}), jsxRuntime.jsx(CaptchaContext.CaptchaContextProvider, { children: jsxRuntime.jsx(AccountExistsContext.AccountExistsContextProvider, { children: jsxRuntime.jsx(WalletContext.WalletContextProvider, { canSync: didConnectedStateLoad, children: jsxRuntime.jsx(PasskeyContext.PasskeyContextProvider, { children: jsxRuntime.jsx(EmailVerificationContext.EmailVerificationContextProvider, { children: jsxRuntime.jsx(DynamicWidgetContext.DynamicWidgetContextProvider, { children: jsxRuntime.jsx(DynamicBridgeWidgetContext.DynamicBridgeWidgetContextProvider, { children: jsxRuntime.jsx(FundingContext.FundingContextProvider, { children: jsxRuntime.jsx(AccessDeniedContext.AccessDeniedContextProvider, { children: jsxRuntime.jsx(SendBalanceContext.SendBalanceContextProvider, { children: jsxRuntime.jsx(WalletGroupContext.WalletGroupContextProvider, { children: jsxRuntime.jsx(UserFieldEditorContext.UserFieldEditorContextProvider, { children: jsxRuntime.jsx(ConnectWithEmailOtpProvider.ConnectWithEmailOtpProvider, { children: jsxRuntime.jsx(PhantomRedirectContext.PhantomRedirectContextProvider, { children: jsxRuntime.jsxs(SocialRedirectContext.SocialRedirectContextProvider, { children: [jsxRuntime.jsx(app.DynamicAuthFlow, {}), jsxRuntime.jsx(Toolkit.Toolkit, {}), jsxRuntime.jsx(FundingWidget.FundingWidget, {}), jsxRuntime.jsx(SyncPasskeyFlow.SyncPasskeyFlow, {}), confirmationModal, children] }) }) }) }) }) }) }) }) }) }) }) }) }) }) })] }) }) }) }) }) }));
};
/** The context provider you need to have access too all of Dynamic's hooks */
const DynamicContextProvider = (props) => {
    useParentDynamicContextGuard.useParentDynamicContextGuard();
    return (jsxRuntime.jsx(ErrorContext.ErrorContextProvider, { children: jsxRuntime.jsx(UserWalletsContext.UserWalletsProvider, { children: jsxRuntime.jsx(InnerDynamicContextProvider, Object.assign({}, props)) }) }));
};

exports.DynamicContext = DynamicContext;
exports.DynamicContextProvider = DynamicContextProvider;
exports.InnerDynamicContextProvider = InnerDynamicContextProvider;
exports.PUBLIC_PROJECT_LIVE_ENVIRONMENT_ID = PUBLIC_PROJECT_LIVE_ENVIRONMENT_ID;
