// @dynamic-labs/sdk-react-core/src/lib/context/DynamicContext/DynamicContext.js
import { __awaiter } from '../../../../_virtual/_tslib.js';
import { jsx, jsxs } from 'react/jsx-runtime';
import { createContext, useState, useRef, useMemo, useEffect, useCallback } from 'react';
import { I18nextProvider } from 'react-i18next';
import { getIconicSpriteUrl } from '@dynamic-labs/iconic';
import { MissingEnvironmentIdError } from '@dynamic-labs/utils';
import { useWalletBookCdn, getWalletBookCdnUrl, WalletBookContextProvider } from '@dynamic-labs/wallet-book';
import { logger as logger$1 } from '@dynamic-labs/wallet-connector-core';
import { DynamicAuthFlow } from '../../app.js';
import '../../events/dynamicEvents.js';
import { decodeJwt } from '../../shared/utils/functions/decodeJwt/decodeJwt.js';
import { ViewContextProvider } from '../ViewContext/ViewContext.js';
import { logger } from '../../shared/logger.js';
import { BRIDGE_ONBOARDING_COMPLETED, AUTH_TOKEN, AUTH_USER, AUTH_MODE, WALLET_PICKER_SEARCH_KEY } from '../../utils/constants/localStorage.js';
import '../../utils/constants/colors.js';
import { DEFAULT_NUMBER_OF_WALLETS_TO_SHOW } from '../../utils/constants/values.js';
import { LocalStorage } from '../../shared/utils/classes/storage/localStorage.js';
import 'viem';
import '@dynamic-labs/sdk-api';
import { useGlobalLoading } from '../../shared/utils/hooks/useGlobalLoading/useGlobalLoading.js';
import { useLocalStorage } from '../../shared/utils/hooks/useLocalStorage/useLocalStorage.js';
import '../../shared/consts/index.js';
import { CaptchaContextProvider } from '../CaptchaContext/CaptchaContext.js';
import { ErrorContextProvider } from '../ErrorContext/ErrorContext.js';
import '@dynamic-labs/multi-wallet';
import { getAuthToken } from '../../utils/functions/getAuthToken/getAuthToken.js';
import 'viem/chains';
import { AccessDeniedContextProvider } from '../AccessDeniedContext/AccessDeniedContext.js';
import { AccountExistsContextProvider } from '../AccountExistsContext/AccountExistsContext.js';
import { sendDynamicProps } from '../../data/api.js';
import { EmailVerificationContextProvider } from '../EmailVerificationContext/EmailVerificationContext.js';
import 'react-dom';
import { useFetchNonce } from '../../utils/hooks/useFetchNonce/useFetchNonce.js';
import { useFetchProjectSettings } from '../../utils/hooks/useFetchProjectSettings/useFetchProjectSettings.js';
import { ThemeContextProvider } from '../ThemeContext/ThemeContext.js';
import { useSetWalletConnectorVerifiedCredentials } from '../../utils/hooks/useSetWalletConnectorVerifiedCredentials/useSetWalletConnectorVerifiedCredentials.js';
import '@dynamic-labs/types';
import 'yup';
import '../MockContext/MockContext.js';
import '../../utils/hooks/useUserUpdateRequest/useUpdateUser/useUpdateUser.js';
import { UserFieldEditorContextProvider } from '../UserFieldEditorContext/UserFieldEditorContext.js';
import { useValidateSession } from '../../utils/hooks/useValidateSession/useValidateSession.js';
import { useWalletConnectorOptions } from '../../utils/hooks/useWalletConnectorOptions/useWalletConnectorOptions.js';
import { useWalletConnectors } from '../../utils/hooks/useWalletConnectors/useWalletConnectors.js';
import { useWalletUiUtils } from '../../utils/hooks/useWalletUiUtils/useWalletUiUtils.js';
import '../../components/Alert/Alert.js';
import '../../components/ShadowDOM/ShadowDOM.js';
import '../../components/IconButton/IconButton.js';
import '../../components/InlineWidget/InlineWidget.js';
import '../../components/IsBrowser/IsBrowser.js';
import '../../components/MenuList/Dropdown/Dropdown.js';
import '../../components/Transition/ZoomTransition/ZoomTransition.js';
import '../../components/Transition/SlideInUpTransition/SlideInUpTransition.js';
import '../../components/Transition/OpacityTransition/OpacityTransition.js';
import '../../components/Popper/Popper/Popper.js';
import '../../components/Popper/PopperContext/PopperContext.js';
import 'react-focus-lock';
import 'qrcode';
import 'formik';
import { Locale } from '../../locale/locale.js';
import '../../components/OverlayCard/OverlayCard.context.js';
import { SyncPasskeyFlow } from '../../components/SyncPasskeyFlow/SyncPasskeyFlow.js';
import { Toolkit } from '../../components/Toolkit/Toolkit.js';
import '../../components/PasskeyCreatedSuccessBanner/PasskeyCreatedSuccessBanner.js';
import { PasskeyContextProvider } from '../PasskeyContext/PasskeyContext.js';
import '../../views/WalletList/WalletList.js';
import '@hcaptcha/react-hcaptcha';
import { LoadingContextProvider } from '../LoadingContext/LoadingContext.js';
import { SocialRedirectContextProvider } from '../SocialRedirectContext/SocialRedirectContext.js';
import { WalletGroupContextProvider } from '../WalletGroupContext/WalletGroupContext.js';
import '../../widgets/DynamicBridgeWidget/views/WalletsView/components/SecondaryWallets/SecondaryWallets.js';
import '../FooterAnimationContext/index.js';
import ApiEndpoint from '../../config/ApiEndpoint.js';
import { getMissingChains } from '../../utils/functions/getMissingChains/getMissingChains.js';
import { parseTheme } from '../../utils/functions/parseTheme/parseTheme.js';
import { useMultiWalletWidgetState } from '../../utils/hooks/multiWallet/useMultiWalletWidgetState/useMultiWalletWidgetState.js';
import { useClearWalletConnectSessions } from '../../utils/hooks/useClearWalletConnectSessions/useClearWalletConnectSessions.js';
import { useConnectWallet } from '../../utils/hooks/useConnectWallet/useConnectWallet.js';
import { useFetchWalletsForChainsMap } from '../../utils/hooks/useFetchWalletsForChainsMap/useFetchWalletsForChainsMap.js';
import { useHandleUnlinkWallet } from '../../utils/hooks/useHandleUnlinkWallet/useHandleUnlinkWallet.js';
import { usePrimaryWalletId } from '../../utils/hooks/usePrimaryWalletId/usePrimaryWalletId.js';
import { useProcessNetworkConfigurations } from '../../utils/hooks/useProcessNetworkConfigurations/useProcessNetworkConfigurations.js';
import { useRpcProviders } from '../../utils/hooks/useRpcProviders/useRpcProviders.js';
import { useWalletEventListeners } from '../../utils/hooks/useWalletEventListeners/useWalletEventListeners.js';
import { DynamicBridgeWidgetContextProvider } from '../../widgets/DynamicBridgeWidget/context/DynamicBridgeWidgetContext/DynamicBridgeWidgetContext.js';
import { DynamicWidgetContextProvider } from '../../widgets/DynamicWidget/context/DynamicWidgetContext.js';
import '../../widgets/DynamicWidget/components/DynamicWidgetCard/DynamicWidgetCard.js';
import '../../components/UserProfile/parts/UserProfileField/components/VerifiedEmailIcon/VerifiedEmailIcon.js';
import '../../widgets/DynamicWidget/components/DynamicWidgetHeader/DynamicWidgetHeader.js';
import '@dynamic-labs/viem-utils';
import '../../views/TransactionConfirmationView/helpers/transactionErrorMessage.js';
import '../../utils/hooks/useTransactionWithGasPrice/useTransactionWithGasPrice.js';
import '../../widgets/DynamicWidget/views/ManagePasskeysWidgetView/PasskeyCard/PasskeyCard.js';
import { FundingWidget } from '../../widgets/FundingWidget/FundingWidget.js';
import { ConnectWithEmailOtpProvider } from '../ConnectWithEmailOtpContext/ConnectWithEmailOtpProvider.js';
import '../ConnectWithEmailOtpContext/connectWithEmailOtpContext.js';
import { FundingContextProvider } from '../FundingContext/FundingContext.js';
import { SendBalanceContextProvider } from '../SendBalanceContext/SendBalanceContext.js';
import { UseNetworkValidation } from '../UseNetworkValidation/UseNetworkValidation.js';
import { UserWalletsProvider } from '../UserWalletsContext/UserWalletsContext.js';
import { WalletContextProvider } from '../WalletContext/WalletContext.js';
import { PhantomRedirectContextProvider } from '../PhantomRedirectContext/PhantomRedirectContext.js';
import { handleDisconnectWallet } from './handleDisconnectWallet/handleDisconnectWallet.js';
import { resolveNetworkValidationMode, initExpirationTime, getInitialView, getDeepLinkPreference } from './helpers/helpers.js';
import { useDisplayOrderState } from './hooks/useDisplayOrderState/useDisplayOrderState.js';
import { useEmailLoginState } from './hooks/useEmailLoginState/useEmailLoginState.js';
import { useHandleLogout } from './hooks/useHandleLogout/useHandleLogout.js';
import { useNameService } from './hooks/useNameService/useNameService.js';
import { useSelectedWalletConnector } from './hooks/useSelectedWalletConnector/useSelectedWalletConnector.js';
import { useShowAuthFlow } from './hooks/useShowAuthFlow/useShowAuthFlow.js';
import { usePasskeySuccessPopup } from './hooks/usePasskeySuccessPopup/usePasskeySuccessPopup.js';
import { useTieCallbacksToEvents } from './hooks/useTieCallbacksToEvents/useTieCallbacksToEvents.js';
import { useVerificationInProgress } from './hooks/useVerificationInProgress/useVerificationInProgress.js';
import { useMultiWallet } from './hooks/useMultiWallet/useMultiWallet.js';
import { useSetShowLinkNewWalletModal } from './hooks/useSetShowLinkNewWalletModal/useSetShowLinkNewWalletModal.js';
import { useCustomerCallbacks } from './useCustomerCallbacks/useCustomerCallbacks.js';
import { validateAuthUser } from './validators/validators.js';
import { useParentDynamicContextGuard } from './hooks/useParentDynamicContextGuard/useParentDynamicContextGuard.js';

const PUBLIC_PROJECT_LIVE_ENVIRONMENT_ID = '2762a57b-faa4-41ce-9f16-abff9300e2c9';
const EMPTY_INITIAL_USER = undefined;
const DynamicContext = createContext(undefined);
/** The context provider itself we only use internally */
const InnerDynamicContextProvider = ({ children, theme, settings, locale, }) => {
    var _a, _b, _c, _d, _e, _f, _g;
    const { accessDeniedMessagePrimary, accessDeniedMessageSecondary, accessDeniedButton, apiBaseUrl, enableConnectOnlyFallback = false, deepLinkPreference: deepLinkPreferenceProp = 'native', bridgeChains, cssOverrides, defaultNumberOfWalletsToShow = DEFAULT_NUMBER_OF_WALLETS_TO_SHOW, flowNetwork, initialAuthenticationMode = 'connect-and-sign', debugError = false, displaySiweStatement = true, newToWeb3WalletChainMap, enableVisitTrackingOnConnectOnly = true, environmentId, localStorageSuffix, eventsCallbacks, evmNetworks, walletsFilter, logLevel = 'WARN', mobileExperience = 'in-app-browser', notInTheListImageUrl, onboardingImageUrl, policiesConsentInnerComponent, customPrivacyPolicy, privacyPolicyUrl, signWithEmailWalletName, socialMediaLinkText, socialMediaIconUrl, socialMediaUrl, customTermsOfServices, termsOfServiceUrl, toolkitEnabled, siweStatement, shadowDOMEnabled = true, walletConnectors: walletConnectorsProp, socialProvidersFilter, showLockedWalletView = false, walletConnectPreferredChains, walletConnectorExtensions, hideEmbeddedWalletTransactionUIs, } = settings;
    const networkValidationMode = resolveNetworkValidationMode({
        bridgeChains: settings.bridgeChains,
        enableForcedNetworkValidation: settings.enableForcedNetworkValidation,
        networkValidationMode: settings.networkValidationMode,
    });
    let { appLogoUrl = 'https://demo.dynamic.xyz/assets/dynamic-logo.svg', appName = 'Dynamic Example', } = settings;
    const { walletConnectV1Bridge } = settings;
    if (!environmentId) {
        throw new MissingEnvironmentIdError();
    }
    logger.setLogLevel(logLevel);
    logger$1.setLogLevel(logLevel);
    if (environmentId === PUBLIC_PROJECT_LIVE_ENVIRONMENT_ID) {
        logger.warn(`WARNING: DYNAMIC is using a test environment ID ${environmentId}. Please sign up on https://app.dynamic.xyz/ to get your production environment ID.`);
    }
    ApiEndpoint.setBaseUrl((_a = process.env.DYNAMIC_API_BASE_URL) !== null && _a !== void 0 ? _a : apiBaseUrl);
    LocalStorage.setSuffix(localStorageSuffix);
    const i18nSDKInstance = Locale.setup(locale);
    const isAuthenticated = Boolean(getAuthToken());
    const isBridgeFlow = Boolean(bridgeChains);
    const deepLinkPreference = getDeepLinkPreference(deepLinkPreferenceProp, isBridgeFlow);
    const [onboardingOnlyJwt, setOnboardingOnlyJwt] = useState(undefined);
    // state to track the full connectivity status of the individual's wallets.
    // Set to true once the predetermined number of wallets is connected.
    const [bridgeOnboardingCompleted, setBridgeOnboardingCompleted, removeBridgeOnboardingCompleted,] = useLocalStorage(BRIDGE_ONBOARDING_COMPLETED, false);
    const [isFullyConnected, setIsFullyConnected] = useState(false);
    // Allows clients to control widget flow without need to click on AccountControl
    const [showDynamicUserProfile, setShowDynamicUserProfile] = useState(false);
    // Alows clients to control the Bridget Widget flow without need to click on Bridge Widget
    const [showBridgeWidget, setShowBridgeWidget] = useState(false);
    const [isSingleWalletAccount, setIsSingleWalletAccount] = useState(false);
    const [user, setUser, removeUser] = useLocalStorage(AUTH_USER, EMPTY_INITIAL_USER, validateAuthUser, [AUTH_TOKEN]);
    const { consumeNonce } = useFetchNonce(environmentId);
    const [qrcodeUri, setQrcodeUri] = useState('');
    const [desktopUri, setDesktopUri] = useState('');
    const [multiWalletWidgetState, setMultiWalletWidgetState, { awaiting_account_switch: [accountSwitchState], awaiting_signature: [awaitingSignatureState], },] = useMultiWalletWidgetState();
    const { projectSettings, isLoading: isProjectSettingsLoading, removeLsSettings, } = useFetchProjectSettings({
        authToken: getAuthToken(),
        dynamicContextProps: { settings },
        environmentId,
    });
    const multiWallet = useMultiWallet({
        isBridgeFlow,
        multiWalletOverride: (_b = settings.overrides) === null || _b === void 0 ? void 0 : _b.multiWallet,
        multiWalletSettings: (_c = projectSettings === null || projectSettings === void 0 ? void 0 : projectSettings.sdk) === null || _c === void 0 ? void 0 : _c.multiWallet,
    });
    // ensures that when user comes back to site and primary wallet is out of sync, we don't show widget popup
    const [showWidgetStatePopup, setShowWidgetStatePopup] = useState(!multiWallet);
    // Used inside useVerifyOnAwaitingSignature. Is legacy and should be removed asap
    const [legacyIsVerifying, setLegacyIsVerifying] = useState(false);
    // Set to true when a customer starts any auth verification process and set to false when it ends
    const [isVerificationInProgress, setIsVerificationInProgress] = useVerificationInProgress();
    const walletBook = useWalletBookCdn();
    const { configurations: serverNetworkConfigurations, handleRemoveLsNetworks, } = useProcessNetworkConfigurations({
        authToken: getAuthToken(),
        environmentId,
        evmNetworks,
        projectSettings,
    });
    const { imageUserInAccessList, imageUserNotInAccessList, displayName, appLogo, } = (projectSettings === null || projectSettings === void 0 ? void 0 : projectSettings.general) || {};
    appLogoUrl = appLogo || appLogoUrl;
    appName = displayName || appName;
    const [loginWithEmail, setLogInWithEmail, resetLoginWithEmail] = useEmailLoginState((projectSettings === null || projectSettings === void 0 ? void 0 : projectSettings.providers) || [], user);
    const [authorizationViewDisplayOrder, setAuthorizationViewDisplayOrder] = useDisplayOrderState(projectSettings);
    const [confirmationModal, walletUiUtils] = useWalletUiUtils({
        appLogoUrl,
        appName,
        getAppOrigin: () => window.location.origin,
        hideEmbeddedWalletTransactionUIs,
    });
    const { walletConnectorOptions } = useWalletConnectorOptions({
        appLogoUrl,
        appName,
        deepLinkPreference,
        flowNetwork,
        mobileExperience,
        networkConfigurations: serverNetworkConfigurations,
        projectSettings,
        walletBook,
        walletConnectPreferredChains,
        walletConnectV1Bridge,
        walletConnectorExtensions,
        walletConnectorsProp,
        walletUiUtils,
    });
    const [authMode, setAuthMode] = useLocalStorage(AUTH_MODE, initialAuthenticationMode);
    const { clearPrimaryWalletId, primaryWalletId, setPrimaryWalletId } = usePrimaryWalletId();
    const isUserAuthenticated = Boolean(authMode === 'connect-and-sign' ? user : isFullyConnected);
    const isRenderingEmbeddedAuthFlow = useRef(false);
    const [showAuthFlow, setShowAuthFlow] = useShowAuthFlow({
        isAuthenticated: isUserAuthenticated,
        isMultiWalletEnabled: multiWallet,
        isRenderingEmbeddedAuthFlow,
        walletConnectorOptions,
    });
    const { connectWallet, removeConnectedWalletsInfo, connectedWalletsInfo, getConnectedWalletById, connectedWallets, disconnectWallet, refreshConnectedWallet, } = useConnectWallet({
        authMode,
        clearPrimaryWalletId,
        enableVisitTrackingOnConnectOnly,
        environmentId,
        isBridgeFlow,
        onBeforeConnectSuccessConfirmation: eventsCallbacks === null || eventsCallbacks === void 0 ? void 0 : eventsCallbacks.onBeforeConnectSuccessConfirmation,
        onDisconnect: (wallet) => handleDisconnectWallet({
            eventsCallbacks,
            wallet,
        }),
        primaryWalletId,
        setIsVerificationInProgress,
        setPrimaryWalletId,
        setShowAuthFlow,
        user,
        walletConnectorOptions,
    });
    const bridgeChainsToConnect = useMemo(() => getMissingChains(bridgeChains, connectedWalletsInfo), [bridgeChains, connectedWalletsInfo]);
    const { selectedWalletConnector, setSelectedWalletConnectorKey } = useSelectedWalletConnector({ walletConnectorOptions });
    //selected wallet to perform an action (become primary, unlink, etc)
    const [selectedWalletWithAction, setSelectedWalletWithAction] = useState(null);
    // until user completes onboarding, jwt is only stored in onboardingOnlyJwt
    const authToken = (_d = getAuthToken()) !== null && _d !== void 0 ? _d : onboardingOnlyJwt;
    useSetWalletConnectorVerifiedCredentials(authToken, walletConnectorOptions);
    const { didConnectedStateLoad, primaryWallet, showQrcodeModal, secondaryWallets, setPrimaryWallet, setShowQrcodeModal, wallets: linkedOrConnectedWallets, } = useWalletConnectors({
        authMode,
        authToken: getAuthToken() || onboardingOnlyJwt,
        canHaveMultipleWalletsConnected: multiWallet,
        connectedWallets,
        multiWalletWidgetState,
        onboardingOnlyJwt,
        primaryWalletId,
        setDesktopUri,
        setMultiWalletWidgetState,
        setPrimaryWalletId,
        setQrcodeUri,
        user,
        walletConnectorOptions,
    });
    useEffect(() => {
        setIsFullyConnected(!(bridgeChainsToConnect === null || bridgeChainsToConnect === void 0 ? void 0 : bridgeChainsToConnect.length) && Boolean(primaryWallet));
    }, [bridgeChainsToConnect, primaryWallet]);
    const walletConnectors = linkedOrConnectedWallets.map((wallet) => wallet.connector);
    const { sdkHasLoaded, setSdkHasLoaded } = useGlobalLoading({
        authMode,
        connectedInfo: connectedWalletsInfo[0],
        connectedWallets,
        primaryWallet,
        projectSettings,
        user,
        walletBook,
    });
    const setShowLinkNewWalletModal = useSetShowLinkNewWalletModal({
        isAuthenticated: isUserAuthenticated,
        isMultiWalletEnabled: multiWallet,
        setShowAuthFlow,
    });
    const { getNameService, removeConnectedNameService } = useNameService({
        authToken,
        currentWallet: primaryWallet !== null && primaryWallet !== void 0 ? primaryWallet : connectedWallets[0],
    });
    const { clearAllWalletConnectSessions } = useClearWalletConnectSessions({
        connectors: walletConnectors,
    });
    const handleLogOut = useHandleLogout({
        bridgeChains,
        clearAllWalletConnectSessions,
        clearPrimaryWalletId,
        connectedWallets,
        environmentId,
        eventsCallbacks,
        handleRemoveLsNetworks,
        initialAuthenticationMode,
        isVerificationInProgress,
        removeBridgeOnboardingCompleted,
        removeConnectedNameService,
        removeConnectedWalletsInfo,
        removeLsSettings,
        removeUser,
        resetLoginWithEmail,
        setAuthMode,
        setIsVerificationInProgress,
        setLegacyIsVerifying,
        setMultiWalletWidgetState,
        setOnboardingOnlyJwt,
        setSdkHasLoaded,
        setShowAuthFlow,
        setShowBridgeWidget,
        setShowDynamicUserProfile,
        user,
        walletConnectors,
    });
    let wcVersion;
    if (projectSettings) {
        wcVersion =
            ((_f = (_e = projectSettings === null || projectSettings === void 0 ? void 0 : projectSettings.sdk) === null || _e === void 0 ? void 0 : _e.walletConnect) === null || _f === void 0 ? void 0 : _f.v2Enabled) && !walletConnectV1Bridge
                ? 2
                : 1;
    }
    const { setHasLoggedOut } = useValidateSession({
        authToken,
        handleLogOut,
        user,
        walletConnectV1Bridge,
        walletConnectors,
        wcVersion,
    });
    // Ties all client callbacks to dynamic's events, such that
    // when our events are triggered, so are the client callbacks
    useTieCallbacksToEvents({
        clientCallbacks: eventsCallbacks,
        connectedWallets,
        setHasLoggedOut,
        user,
    });
    const handleUnlinkWallet = useHandleUnlinkWallet({
        environmentId,
        eventsCallbacks,
        primaryWalletId,
        secondaryWallets,
        setUser,
        user,
        verifiedCredentials: (user === null || user === void 0 ? void 0 : user.verifiedCredentials) || [],
    });
    const { loadingNetwork, network, setNetwork } = useWalletEventListeners({
        authMode,
        disconnectWallet,
        handleLogOut,
        multiWallet,
        multiWalletWidgetState,
        primaryWallet,
        refreshConnectedWallet,
        secondaryWallets,
        selectedWalletConnector,
        selectedWalletWithAction,
        setMultiWalletWidgetState,
        setPrimaryWalletId,
        setSelectedWalletConnectorKey,
        setSelectedWalletWithAction,
        user,
    });
    const { walletsForChainsMap } = useFetchWalletsForChainsMap(newToWeb3WalletChainMap);
    const rpcProviders = useRpcProviders({
        networkConfigurations: serverNetworkConfigurations,
    });
    useEffect(() => {
        initExpirationTime(handleLogOut);
    }, [handleLogOut]);
    const { setCallback } = useCustomerCallbacks({
        callbacks: {
            onAuthSuccess: eventsCallbacks === null || eventsCallbacks === void 0 ? void 0 : eventsCallbacks.onAuthSuccess,
            onEmbeddedWalletCreated: eventsCallbacks === null || eventsCallbacks === void 0 ? void 0 : eventsCallbacks.onEmbeddedWalletCreated,
            onLinkSuccess: eventsCallbacks === null || eventsCallbacks === void 0 ? void 0 : eventsCallbacks.onLinkSuccess,
            onUserProfileUpdate: eventsCallbacks === null || eventsCallbacks === void 0 ? void 0 : eventsCallbacks.onUserProfileUpdate,
        },
        handleLogOut,
        isAuthenticated,
        primaryWallet,
        secondaryWallets,
        setHasLoggedOut,
        setIsVerificationInProgress,
        user,
    });
    const sendWagmiSettings = useCallback(() => (settings) => {
        const { dynamicWagmiSettings } = settings;
        if (dynamicWagmiSettings) {
            sendDynamicProps(environmentId, {
                dynamicWagmiSettings: { dynamicWagmiSettings },
            });
        }
    }, [environmentId]);
    const clearStatesOnBackClick = useCallback(() => __awaiter(void 0, void 0, void 0, function* () {
        setDesktopUri('');
        // Clear verification flag but dont require it to be set
        if (isVerificationInProgress)
            setIsVerificationInProgress(false);
        setLegacyIsVerifying(false);
        if (!user) {
            clearAllWalletConnectSessions();
        }
        setMultiWalletWidgetState('idle');
        if (authMode !== 'connect-only') {
            removeConnectedWalletsInfo();
        }
        if (!LocalStorage.getFromLS(AUTH_USER)) {
            LocalStorage.removeFromLS(AUTH_TOKEN);
        }
        if (LocalStorage.getFromLS(WALLET_PICKER_SEARCH_KEY)) {
            LocalStorage.removeFromLS(WALLET_PICKER_SEARCH_KEY);
        }
        if (selectedWalletConnector &&
            (!primaryWallet ||
                primaryWallet.connector.key !== selectedWalletConnector.key)) {
            yield (selectedWalletConnector === null || selectedWalletConnector === void 0 ? void 0 : selectedWalletConnector.endSession());
        }
    }), [
        isVerificationInProgress,
        setIsVerificationInProgress,
        user,
        setMultiWalletWidgetState,
        authMode,
        selectedWalletConnector,
        primaryWallet,
        clearAllWalletConnectSessions,
        removeConnectedWalletsInfo,
    ]);
    const { passkeySuccessPopup, setPasskeySuccessPopup } = usePasskeySuccessPopup();
    const value = useMemo(() => {
        var _a, _b;
        return ({
            accessDeniedButton,
            accessDeniedMessagePrimary,
            accessDeniedMessageSecondary,
            accountSwitchState,
            appLogoUrl,
            appName,
            authMode,
            authToken: getAuthToken(),
            authorizationViewDisplayOrder,
            awaitingSignatureState,
            bridgeChains,
            bridgeChainsToConnect,
            bridgeOnboardingCompleted,
            clearStatesOnBackClick,
            connectWallet,
            connectedWallets,
            consumeNonce,
            cssOverrides,
            customPrivacyPolicy,
            customTermsOfServices,
            debugError,
            decodedAuthToken: decodeJwt(getAuthToken()),
            decodedOnboardingOnlyToken: decodeJwt(onboardingOnlyJwt),
            defaultNumberOfWalletsToShow,
            desktopUri,
            disconnectWallet,
            displaySiweStatement,
            enableConnectOnlyFallback,
            environmentId,
            eventsCallbacks,
            getConnectedWalletById,
            getNameService,
            handleLogOut,
            handleUnlinkWallet: authMode === 'connect-only' ? disconnectWallet : handleUnlinkWallet,
            i18nSDKInstance,
            isAuthenticated,
            isBridgeFlow,
            isFullyConnected,
            isProjectSettingsLoading,
            isRenderingEmbeddedAuthFlow,
            isSingleWalletAccount,
            isVerificationInProgress,
            legacyIsVerifying,
            linkedWallets: authMode === 'connect-only' || !user ? [] : linkedOrConnectedWallets,
            loadingNetwork,
            locale: {
                changeLanguage: Locale.getInstance().changeLanguage,
            },
            loginWithEmail,
            multiWallet,
            multiWalletWidgetState,
            network,
            networkConfigurations: serverNetworkConfigurations,
            networkValidationMode,
            newToWeb3WalletChainMap: walletsForChainsMap,
            notInTheListImageUrl: imageUserNotInAccessList || notInTheListImageUrl,
            onboardingImageUrl: imageUserInAccessList || onboardingImageUrl,
            onboardingOnlyJwt,
            overrides: settings.overrides,
            passkeySuccessPopup,
            policiesConsentInnerComponent,
            primaryWallet,
            primaryWalletId,
            privacyPolicyUrl,
            projectSettings,
            qrcodeUri,
            removeBridgeOnboardingCompleted,
            removeConnectedWalletsInfo,
            rpcProviders,
            sdkHasLoaded,
            secondaryWallets,
            selectedWalletConnector,
            selectedWalletWithAction,
            sendWagmiSettings,
            setAuthMode,
            setAuthorizationViewDisplayOrder,
            setBridgeOnboardingCompleted,
            setCallback,
            setDesktopUri,
            setIsSingleWalletAccount,
            setIsVerificationInProgress,
            setLegacyIsVerifying,
            setLogInWithEmail,
            setMultiWalletWidgetState,
            setNetwork,
            setOnboardingOnlyJwt,
            setPasskeySuccessPopup,
            setPrimaryWallet,
            setPrimaryWalletId,
            setQrcodeUri,
            setSdkHasLoaded,
            setSelectedWalletConnectorKey,
            setSelectedWalletWithAction,
            setShowAuthFlow,
            setShowBridgeWidget,
            setShowDynamicUserProfile,
            setShowLinkNewWalletModal,
            setShowQrcodeModal,
            setShowWidgetStatePopup,
            setUser,
            shadowDOMEnabled,
            showAuthFlow,
            showBridgeWidget,
            showDynamicUserProfile,
            showLockedWalletView,
            showQrcodeModal,
            showWidgetStatePopup,
            signWithEmailWalletName,
            siweStatement,
            socialMediaIconUrl: ((_a = projectSettings === null || projectSettings === void 0 ? void 0 : projectSettings.general) === null || _a === void 0 ? void 0 : _a.supportUrls) || socialMediaIconUrl,
            socialMediaLinkText,
            socialMediaUrl,
            socialProvidersFilter,
            termsOfServiceUrl,
            theme,
            toolkitEnabled,
            user,
            walletConnector: (_b = primaryWallet === null || primaryWallet === void 0 ? void 0 : primaryWallet.connector) !== null && _b !== void 0 ? _b : null,
            walletConnectorOptions,
            walletUiUtils,
            walletsFilter,
        });
    }, [
        accessDeniedButton,
        accessDeniedMessagePrimary,
        accessDeniedMessageSecondary,
        accountSwitchState,
        appLogoUrl,
        appName,
        authMode,
        authorizationViewDisplayOrder,
        awaitingSignatureState,
        bridgeChains,
        bridgeChainsToConnect,
        bridgeOnboardingCompleted,
        clearStatesOnBackClick,
        connectWallet,
        connectedWallets,
        consumeNonce,
        cssOverrides,
        customPrivacyPolicy,
        customTermsOfServices,
        debugError,
        onboardingOnlyJwt,
        defaultNumberOfWalletsToShow,
        desktopUri,
        disconnectWallet,
        displaySiweStatement,
        enableConnectOnlyFallback,
        environmentId,
        eventsCallbacks,
        getConnectedWalletById,
        getNameService,
        handleLogOut,
        handleUnlinkWallet,
        i18nSDKInstance,
        isAuthenticated,
        isBridgeFlow,
        isFullyConnected,
        isProjectSettingsLoading,
        isSingleWalletAccount,
        isVerificationInProgress,
        legacyIsVerifying,
        user,
        linkedOrConnectedWallets,
        loadingNetwork,
        loginWithEmail,
        multiWallet,
        multiWalletWidgetState,
        network,
        serverNetworkConfigurations,
        networkValidationMode,
        walletsForChainsMap,
        imageUserNotInAccessList,
        notInTheListImageUrl,
        imageUserInAccessList,
        onboardingImageUrl,
        settings.overrides,
        passkeySuccessPopup,
        policiesConsentInnerComponent,
        primaryWallet,
        primaryWalletId,
        privacyPolicyUrl,
        projectSettings,
        qrcodeUri,
        removeBridgeOnboardingCompleted,
        removeConnectedWalletsInfo,
        rpcProviders,
        sdkHasLoaded,
        secondaryWallets,
        selectedWalletConnector,
        selectedWalletWithAction,
        sendWagmiSettings,
        setAuthMode,
        setAuthorizationViewDisplayOrder,
        setBridgeOnboardingCompleted,
        setCallback,
        setIsVerificationInProgress,
        setLogInWithEmail,
        setMultiWalletWidgetState,
        setNetwork,
        setPrimaryWallet,
        setPrimaryWalletId,
        setSdkHasLoaded,
        setSelectedWalletConnectorKey,
        setShowAuthFlow,
        setShowLinkNewWalletModal,
        setPasskeySuccessPopup,
        setShowQrcodeModal,
        setUser,
        shadowDOMEnabled,
        showAuthFlow,
        showBridgeWidget,
        showDynamicUserProfile,
        showLockedWalletView,
        showQrcodeModal,
        showWidgetStatePopup,
        signWithEmailWalletName,
        siweStatement,
        socialMediaIconUrl,
        socialMediaLinkText,
        socialMediaUrl,
        socialProvidersFilter,
        termsOfServiceUrl,
        theme,
        toolkitEnabled,
        walletConnectorOptions,
        walletUiUtils,
        walletsFilter,
    ]);
    useEffect(() => {
        const addPrefetch = (attr) => {
            if (!document.head.querySelector(`#${attr.id}`)) {
                const element = document.createElement('link');
                element.setAttribute('id', attr.id);
                element.setAttribute('rel', 'prefetch');
                element.setAttribute('href', attr.href);
                element.setAttribute('as', attr.as);
                element.setAttribute('type', attr.type);
                document.head.insertAdjacentElement('beforeend', element);
            }
        };
        addPrefetch({
            as: 'image',
            href: getIconicSpriteUrl(),
            id: 'sprite',
            type: 'image/svg+xml',
        });
        addPrefetch({
            as: 'fetch',
            href: getWalletBookCdnUrl(),
            id: 'wallet-book',
            type: 'application/json',
        });
    }, []);
    // DYN-1140 - Opens Widget when multiWalletWidgetState changes
    useEffect(() => {
        // If statement to secure SingleWallet widget
        if (!multiWallet && multiWalletWidgetState !== 'awaiting_account_switch') {
            return;
        }
        // We need to make sure that the user is logged in so that we do not try to display the widget before authorization
        if (user && multiWalletWidgetState !== 'idle' && !showDynamicUserProfile) {
            setShowDynamicUserProfile(true);
        }
    }, [multiWallet, multiWalletWidgetState, showDynamicUserProfile, user]);
    return (jsx(I18nextProvider, { i18n: i18nSDKInstance, children: jsx(DynamicContext.Provider, { value: value, children: jsx(WalletBookContextProvider, { walletBook: walletBook, children: jsx(ThemeContextProvider, { customerTheme: parseTheme(theme, ((_g = projectSettings === null || projectSettings === void 0 ? void 0 : projectSettings.design) === null || _g === void 0 ? void 0 : _g.modal) || undefined), designSettings: projectSettings === null || projectSettings === void 0 ? void 0 : projectSettings.design, children: jsx(LoadingContextProvider, { children: jsxs(ViewContextProvider, { initialView: getInitialView({
                                bridgeOnboardingCompleted,
                                connectedWallets,
                                isAuthenticated: isAuthenticated || Boolean(primaryWallet),
                                isBridgeFlow,
                                isMultiWalletEnabled: multiWallet,
                            }), children: [networkValidationMode === 'always' && jsx(UseNetworkValidation, {}), jsx(CaptchaContextProvider, { children: jsx(AccountExistsContextProvider, { children: jsx(WalletContextProvider, { canSync: didConnectedStateLoad, children: jsx(PasskeyContextProvider, { children: jsx(EmailVerificationContextProvider, { children: jsx(DynamicWidgetContextProvider, { children: jsx(DynamicBridgeWidgetContextProvider, { children: jsx(FundingContextProvider, { children: jsx(AccessDeniedContextProvider, { children: jsx(SendBalanceContextProvider, { children: jsx(WalletGroupContextProvider, { children: jsx(UserFieldEditorContextProvider, { children: jsx(ConnectWithEmailOtpProvider, { children: jsx(PhantomRedirectContextProvider, { children: jsxs(SocialRedirectContextProvider, { children: [jsx(DynamicAuthFlow, {}), jsx(Toolkit, {}), jsx(FundingWidget, {}), jsx(SyncPasskeyFlow, {}), confirmationModal, children] }) }) }) }) }) }) }) }) }) }) }) }) }) }) })] }) }) }) }) }) }));
};
/** The context provider you need to have access too all of Dynamic's hooks */
const DynamicContextProvider = (props) => {
    useParentDynamicContextGuard();
    return (jsx(ErrorContextProvider, { children: jsx(UserWalletsProvider, { children: jsx(InnerDynamicContextProvider, Object.assign({}, props)) }) }));
};

export { DynamicContext, DynamicContextProvider, InnerDynamicContextProvider, PUBLIC_PROJECT_LIVE_ENVIRONMENT_ID };
