// @zerodev/sdk/dist/cjs/kernel-zerodev/utils/calc-pre-verification-gas.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.calcPreVerificationGas = exports.DefaultGasOverheads = void 0;
const viem_1 = require("viem");
const ERC4337_utils_js_1 = require("./ERC4337-utils.js");
exports.DefaultGasOverheads = {
    fixed: 21000,
    perUserOp: 18300,
    perUserOpWord: 4,
    zeroByte: 4,
    nonZeroByte: 16,
    bundleSize: 1,
    sigSize: 65,
};
function calcPreVerificationGas(userOp, overheads) {
    const ov = { ...exports.DefaultGasOverheads, ...(overheads ?? {}) };
    const p = {
        preVerificationGas: 21000,
        signature: (0, viem_1.toHex)(Buffer.alloc(ov.sigSize, 1)),
        ...userOp,
    };
    const packed = (0, viem_1.toBytes)((0, ERC4337_utils_js_1.packUserOp)(p, false));
    const callDataCost = packed
        .map((x) => (x === 0 ? ov.zeroByte : ov.nonZeroByte))
        .reduce((sum, x) => sum + x);
    const ret = Math.round(callDataCost +
        ov.fixed / ov.bundleSize +
        ov.perUserOp +
        ov.perUserOpWord * packed.length);
    return ret;
}
exports.calcPreVerificationGas = calcPreVerificationGas;
