// @zerodev/sdk/dist/cjs/kernel-zerodev/validator/erc165-session-key-validator.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ERC165SessionKeyValidator = void 0;
const aa_core_1 = require("@alchemy/aa-core");
const base_js_1 = require("./base.js");
const viem_1 = require("viem");
const ERC165SessionKeyValidatorAbi_js_1 = require("../abis/ERC165SessionKeyValidatorAbi.js");
const constants_js_1 = require("../constants.js");
const KernelAccountAbi_js_1 = require("../abis/KernelAccountAbi.js");
const index_js_1 = require("../api/index.js");
const utils_js_1 = require("../utils.js");
class ERC165SessionKeyValidator extends base_js_1.KernelBaseValidator {
    constructor(params) {
        super(params);
        Object.defineProperty(this, "sessionKey", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: void 0
        });
        Object.defineProperty(this, "sessionKeyData", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: void 0
        });
        this.sessionKey = params.sessionKey;
        this.sessionKeyData = params.sessionKeyData;
        this.mode = params.mode ?? base_js_1.ValidatorMode.plugin;
    }
    static async init(params) {
        const chainId = await (0, index_js_1.getChainId)(params.projectId);
        if (!chainId) {
            throw new Error("ChainId not found");
        }
        const chain = (0, utils_js_1.getChain)(chainId);
        const instance = new ERC165SessionKeyValidator({ ...params, chain });
        return instance;
    }
    async signer() {
        return await Promise.resolve(this.sessionKey);
    }
    async getEnableData() {
        return (0, viem_1.concat)([
            await this.sessionKey.getAddress(),
            (0, viem_1.pad)(this.sessionKeyData.erc165InterfaceId, { size: 4 }),
            (0, viem_1.pad)(this.sessionKeyData.selector, { size: 4 }),
            (0, viem_1.pad)((0, viem_1.toHex)(this.sessionKeyData.validUntil), { size: 6 }),
            (0, viem_1.pad)((0, viem_1.toHex)(this.sessionKeyData.validAfter), { size: 6 }),
            (0, viem_1.pad)((0, viem_1.toHex)(this.sessionKeyData.addressOffset), { size: 4 }),
        ]);
    }
    async isPluginEnabled(kernelAccountAddress, selector) {
        if (!this.publicClient) {
            throw new Error("Validator uninitialized: PublicClient missing");
        }
        const execDetail = await this.publicClient.readContract({
            abi: KernelAccountAbi_js_1.KernelAccountAbi,
            address: kernelAccountAddress,
            functionName: "getExecution",
            args: [selector],
        });
        const enableData = await this.publicClient.readContract({
            abi: ERC165SessionKeyValidatorAbi_js_1.ERC165SessionKeyValidatorAbi,
            address: this.validatorAddress,
            functionName: "sessionKeys",
            args: [await this.sessionKey.getAddress(), kernelAccountAddress],
        });
        const enableDataHex = (0, viem_1.concat)([
            await this.sessionKey.getAddress(),
            (0, viem_1.pad)(enableData[2], { size: 4 }),
            (0, viem_1.pad)(enableData[1], { size: 4 }),
            (0, viem_1.pad)((0, viem_1.toHex)(enableData[3]), { size: 6 }),
            (0, viem_1.pad)((0, viem_1.toHex)(enableData[4]), { size: 6 }),
            (0, viem_1.pad)((0, viem_1.toHex)(enableData[5]), { size: 4 }),
        ]);
        return (execDetail.validator.toLowerCase() ===
            this.validatorAddress.toLowerCase() &&
            enableData[0] &&
            enableDataHex.toLowerCase() === (await this.getEnableData()).toLowerCase());
    }
    async getDummyUserOpSignature() {
        return constants_js_1.DUMMY_ECDSA_SIG;
    }
    encodeEnable(sessionKeyEnableData) {
        return (0, viem_1.encodeFunctionData)({
            abi: ERC165SessionKeyValidatorAbi_js_1.ERC165SessionKeyValidatorAbi,
            functionName: "enable",
            args: [sessionKeyEnableData],
        });
    }
    encodeDisable(sessionKey) {
        return (0, viem_1.encodeFunctionData)({
            abi: ERC165SessionKeyValidatorAbi_js_1.ERC165SessionKeyValidatorAbi,
            functionName: "disable",
            args: [sessionKey],
        });
    }
    async signMessage(message) {
        return await this.sessionKey.signMessage(message);
    }
    async signTypedData(params) {
        return (0, utils_js_1.fixSignedData)(await this.sessionKey.signTypedData(params));
    }
    async signUserOp(userOp) {
        if (!this.chain) {
            throw new Error("Validator uninitialized");
        }
        const hash = (0, aa_core_1.getUserOperationHash)({
            ...userOp,
            signature: "0x",
        }, this.entryPointAddress, BigInt(this.chain.id));
        const formattedMessage = typeof hash === "string" ? (0, viem_1.toBytes)(hash) : hash;
        return await this.sessionKey.signMessage(formattedMessage);
    }
}
exports.ERC165SessionKeyValidator = ERC165SessionKeyValidator;
