// @zerodev/sdk/dist/cjs/kernel-zerodev/validator/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.RecoveryValidator = exports.SessionKeyValidator = exports.ERC165SessionKeyValidator = exports.KillSwitchValidator = exports.ECDSAValidator = exports.Validators = void 0;
const ecdsa_validator_js_1 = require("./ecdsa-validator.js");
Object.defineProperty(exports, "ECDSAValidator", { enumerable: true, get: function () { return ecdsa_validator_js_1.ECDSAValidator; } });
const erc165_session_key_validator_js_1 = require("./erc165-session-key-validator.js");
Object.defineProperty(exports, "ERC165SessionKeyValidator", { enumerable: true, get: function () { return erc165_session_key_validator_js_1.ERC165SessionKeyValidator; } });
const kill_switch_validator_js_1 = require("./kill-switch-validator.js");
Object.defineProperty(exports, "KillSwitchValidator", { enumerable: true, get: function () { return kill_switch_validator_js_1.KillSwitchValidator; } });
const recovery_validator_js_1 = require("./recovery-validator.js");
Object.defineProperty(exports, "RecoveryValidator", { enumerable: true, get: function () { return recovery_validator_js_1.RecoveryValidator; } });
const session_key_validator_js_1 = require("./session-key-validator.js");
Object.defineProperty(exports, "SessionKeyValidator", { enumerable: true, get: function () { return session_key_validator_js_1.SessionKeyValidator; } });
exports.Validators = {
    ECDSA: ecdsa_validator_js_1.ECDSAValidator,
    KILL_SWITCH: kill_switch_validator_js_1.KillSwitchValidator,
    ERC165_SESSION_KEY: erc165_session_key_validator_js_1.ERC165SessionKeyValidator,
    SESSION_KEY: session_key_validator_js_1.SessionKeyValidator,
    RECOVERY: recovery_validator_js_1.RecoveryValidator,
};
