// @dynamic-labs/sdk-react-core/src/lib/utils/constants/values.cjs
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var sdkApi = require('@dynamic-labs/sdk-api');

const DEFAULT_NUMBER_OF_WALLETS_TO_SHOW = 10;
// references:
// https://www.iana.org/assignments/jwt/jwt.xhtml#claims
// https://github.com/jaredhanson/id-blockchain-claims-in-jwt/blob/main/draft.md
const ChainEnumToVerifiedCredentialName = {
    ALGO: 'algorand',
    BTC: 'bip122', // see: https://namespaces.chainagnostic.org/bip122/caip2
    COSMOS: 'cosmos',
    ETH: 'eip155',
    EVM: 'eip155',
    FLOW: 'flow',
    SOL: 'solana',
    STARK: 'starknet',
};
const VerifiedCredentialNameToChainEnum = {
    algorand: sdkApi.ChainEnum.Algo,
    bip122: sdkApi.ChainEnum.Btc,
    cosmos: sdkApi.ChainEnum.Cosmos,
    eip155: sdkApi.ChainEnum.Evm,
    flow: sdkApi.ChainEnum.Flow,
    solana: sdkApi.ChainEnum.Sol,
    starknet: sdkApi.ChainEnum.Stark,
};

exports.ChainEnumToVerifiedCredentialName = ChainEnumToVerifiedCredentialName;
exports.DEFAULT_NUMBER_OF_WALLETS_TO_SHOW = DEFAULT_NUMBER_OF_WALLETS_TO_SHOW;
exports.VerifiedCredentialNameToChainEnum = VerifiedCredentialNameToChainEnum;
